{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "abXH1EWELa8o"
   },
   "source": [
    "# Topic Modelling\n",
    "\n",
    "Contains topic modeling with BERTopic using OpenAI GPT-4o-mini and pre-computed PubMedBERT and tSNE embeddings."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "dvduAcVTLtgY"
   },
   "source": [
    "## ✋Set Up"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "y7jjL_5LL2_N",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Set up GPUs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 437,
     "status": "ok",
     "timestamp": 1728493661222,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "upmt3RukLtAW",
    "outputId": "3e98e556-ddad-4407-9efa-6b5037536b1b"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Oct  9 17:07:40 2024       \n",
      "+---------------------------------------------------------------------------------------+\n",
      "| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |\n",
      "|-----------------------------------------+----------------------+----------------------+\n",
      "| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |\n",
      "| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |\n",
      "|                                         |                      |               MIG M. |\n",
      "|=========================================+======================+======================|\n",
      "|   0  Tesla T4                       Off | 00000000:00:04.0 Off |                    0 |\n",
      "| N/A   39C    P8              11W /  70W |      0MiB / 15360MiB |      0%      Default |\n",
      "|                                         |                      |                  N/A |\n",
      "+-----------------------------------------+----------------------+----------------------+\n",
      "                                                                                         \n",
      "+---------------------------------------------------------------------------------------+\n",
      "| Processes:                                                                            |\n",
      "|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |\n",
      "|        ID   ID                                                             Usage      |\n",
      "|=======================================================================================|\n",
      "|  No running processes found                                                           |\n",
      "+---------------------------------------------------------------------------------------+\n"
     ]
    }
   ],
   "source": [
    "# GPU information:\n",
    "\n",
    "gpu_info = !nvidia-smi\n",
    "gpu_info = '\\n'.join(gpu_info)\n",
    "if gpu_info.find('failed') >= 0:\n",
    "  print('Not connected to a GPU')\n",
    "else:\n",
    "  print(gpu_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "a2eMQbdKMO56"
   },
   "source": [
    "In order to use a GPU with your notebook, select the **Runtime > Change runtime** type menu, and then set the hardware accelerator dropdown to GPU."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "yxhAcmuGMYYW",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### High RAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 2,
     "status": "ok",
     "timestamp": 1728493661692,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "OjgrnJVQMOZS",
    "outputId": "e102c23d-a0e6-44dd-94ed-ebf3fbf365f5"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your runtime has 54.8 gigabytes of available RAM\n",
      "\n",
      "You are using a high-RAM runtime!\n"
     ]
    }
   ],
   "source": [
    "from psutil import virtual_memory\n",
    "ram_gb = virtual_memory().total / 1e9\n",
    "print('Your runtime has {:.1f} gigabytes of available RAM\\n'.format(ram_gb))\n",
    "\n",
    "if ram_gb < 20:\n",
    "  print('Not using a high-RAM runtime')\n",
    "else:\n",
    "  print('You are using a high-RAM runtime!')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "nqjDwsLJMh1U"
   },
   "source": [
    "Users who have purchased one of Colab's paid plans have access to high-memory VMs when they are available.\n",
    "\n",
    "You can see how much memory you have available at any time by running the following code cell. If the execution result of running the code cell below is \"Not using a high-RAM runtime\", then you can enable a high-RAM runtime via **Runtime > Change runtime** type in the menu. Then select High-RAM in the Runtime shape dropdown. After, re-execute the code cell."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "bMLNM2O_NCcO",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Install libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 154458,
     "status": "ok",
     "timestamp": 1728493833979,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "9jP_pK6OsjmM",
    "outputId": "c28a6299-180f-4ede-82e0-e67a91a97b5d"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m44.4/44.4 kB\u001b[0m \u001b[31m3.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m143.5/143.5 kB\u001b[0m \u001b[31m12.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m9.6/9.6 MB\u001b[0m \u001b[31m101.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m670.2/670.2 MB\u001b[0m \u001b[31m2.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m6.9/6.9 MB\u001b[0m \u001b[31m105.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m9.9/9.9 MB\u001b[0m \u001b[31m60.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m122.4/122.4 MB\u001b[0m \u001b[31m18.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m383.7/383.7 kB\u001b[0m \u001b[31m28.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m410.6/410.6 MB\u001b[0m \u001b[31m3.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m14.1/14.1 MB\u001b[0m \u001b[31m108.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m23.7/23.7 MB\u001b[0m \u001b[31m90.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m823.6/823.6 kB\u001b[0m \u001b[31m50.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m731.7/731.7 MB\u001b[0m \u001b[31m2.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m121.6/121.6 MB\u001b[0m \u001b[31m18.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.5/56.5 MB\u001b[0m \u001b[31m37.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m124.2/124.2 MB\u001b[0m \u001b[31m6.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m196.0/196.0 MB\u001b[0m \u001b[31m4.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m209.8/209.8 MB\u001b[0m \u001b[31m4.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m99.1/99.1 kB\u001b[0m \u001b[31m10.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m89.2/89.2 MB\u001b[0m \u001b[31m24.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.2/4.2 MB\u001b[0m \u001b[31m102.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m76.4/76.4 kB\u001b[0m \u001b[31m5.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m78.0/78.0 kB\u001b[0m \u001b[31m6.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m325.2/325.2 kB\u001b[0m \u001b[31m26.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m245.3/245.3 kB\u001b[0m \u001b[31m19.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.9/2.9 MB\u001b[0m \u001b[31m87.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m85.7/85.7 kB\u001b[0m \u001b[31m8.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.9/56.9 kB\u001b[0m \u001b[31m5.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m58.3/58.3 kB\u001b[0m \u001b[31m5.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25h\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "torchaudio 2.4.1+cu121 requires torch==2.4.1, but you have torch 2.1.0 which is incompatible.\u001b[0m\u001b[31m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#installing for this work.\n",
    "!pip install --quiet  bertopic==0.16.3 scikit-learn==1.2.2 torch==2.1.0 torchvision==0.16 transformers==4.45.1 bitsandbytes==0.44.1 openai==1.51.1 #openai accelerate bitsandbytes xformers adjustText tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "I7eIusVQ-dQj"
   },
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "q3V4lhBQMOZE",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 📄 Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VQ54sZqRNwUc"
   },
   "source": [
    "Import datasets, that contains full embeddings and reduced embeddings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "Kh8LOiENNLFV"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_hdf('Files/embeddings_full_tSNE_uMAP_01MAR2024.h5', key='embeddings')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 372
    },
    "executionInfo": {
     "elapsed": 1242,
     "status": "ok",
     "timestamp": 1728493850640,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "vOs_3EiaNLCv",
    "outputId": "6ecc7e0c-dc1d-4073-b198-721a95201191"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "summary": "{\n  \"name\": \"df\",\n  \"rows\": 80656,\n  \"fields\": [\n    {\n      \"column\": \"pmid\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 5237503,\n        \"min\": 9748443,\n        \"max\": 38377545,\n        \"num_unique_values\": 80656,\n        \"samples\": [\n          33463536,\n          31434796,\n          36445079\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"title\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 80599,\n        \"samples\": [\n          \"Responsive MRI agents for sensing metabolism in vivo.\",\n          \"Metabolomics and Machine Learning Identify Metabolic Differences and Potential Biomarkers for Frequent Versus Infrequent Gout Flares.\",\n          \"Postprandial NMR-Based Metabolic Exchanges Reflect Impaired Phenotypic Flexibility across Splanchnic Organs in the Obese Yucatan Mini-Pig.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"abstract\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 80583,\n        \"samples\": [\n          \"Drought is a misfortune for agriculture and human beings. The annual crop yield reduction caused by drought exceeds the sum of all pathogens. As one of the gatekeepers of China's \\\"granary\\\", rice is the most important to reveal the key drought tolerance factors in rice. Rice seedlings of Nipponbare (Oryza sativa L. ssp. Japonica) were subjected to simulated drought stress, and their root systems were analyzed for the non-targeted metabolome and strand-specific transcriptome. We found that both DEGs and metabolites were enriched in purine metabolism, and allantoin accumulated significantly in roots under drought stress. However, few studies on drought tolerance of exogenous allantoin in rice have been reported. We aimed to further determine whether allantoin can improve the drought tolerance of rice. Under the treatment of exogenous allantoin at different concentrations, the drought resistant metabolites of plants accumulated significantly, including proline and soluble sugar, and reactive oxygen species (ROS) decreased and reached a significant level in 100 \\u03bcmol L[-1]. To this end, a follow-up study was identified in 100 \\u03bcmol L[-1] exogenous allantoin and found that exogenous allantoin improved the drought resistance of rice. At the gene level, under allantoin drought treatment, we found that genes of scavenge reactive oxygen species were significantly expressed, including peroxidase (POD), catalase (CATA), ascorbate peroxidase 8 (APX8) and respiratory burst oxidase homolog protein F (RbohF). This indicates that plants treated by allantoin have better ability to scavenge reactive oxygen species to resist drought. Alternative splicing analysis revealed a total of 427 differentially expressed alternative splicing events across 320 genes. The analysis of splicing factors showed that gene alternative splicing could be divided into many different subgroups and play a regulatory role in many aspects. Through further analysis, we restated the key genes and enzymes in the allantoin synthesis and catabolism pathway, and found that the expression of synthetase and hydrolase showed a downward trend. The pathway of uric acid to allantoin is completed by uric acid oxidase (UOX). To find out the key transcription factors that regulate the expression of this gene, we identified two highly related transcription factors OsERF059 and ONAC007 through correlation analysis. They may be the key for allantoin to enhance the drought resistance of rice.\",\n          \"The study aim was to evaluate the effects of public lockdown during the covid-19 pandemic on glucose and metabolic parameters as well as body weight control in type 2 diabetic patients.\",\n          \"1. Geniposide (genipin 1-O-glucose), one of the major bioactive constituents isolated from Fructus Gardeniae, possesses many biological activities. In this study, an efficient strategy was developed using ultra-high-performance liquid chromatography coupled with linear ion trap-Orbitrap mass spectrometer (UPLC-LTQ-Orbitrap) to profile the in vitro and in vivo metabolic patterns of geniposide in rat liver microsomes (RLMs), plasma, urine, and various tissues. And post-acquisition data-mining methods including extracted ion chromatogram (EIC), multiple mass defect filters (MMDF), fragment ion searching (FISh), and isotope pattern filtering (IPF) were adopted to characterize the known and unknown metabolites. 2. A total of 33 metabolites were detected and interpreted according to accurate mass measurement, diagnostic fragment ions, relevant drug biotransformation knowledge, and bibliography data. Among them, 17 metabolites were detected in the plasma, 31 metabolites were identified in the urine, six metabolites could be found in rat heart, 12 in liver, three in spleen, six in lung, 12 in kidney, six in brain, and four in RLMs. 3. A series of corresponding reactions such as hydrolysis, hydroxylation, taurine conjugation, hydrogenation, decarboxylation, demethylation, sulfate conjugation, cysteine S-conjugation, glucosylation, and their composite reactions were all discovered. 4. The results could provide comprehensive insights and guidance for elucidation of side effect mechanism and safety monitoring as well as for rational formulation design in drug delivery system. The newly discovered geniposide metabolites could be targets for future metabolism studies on the important chemical constituents from herbal medicines.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"language\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 17,\n        \"samples\": [\n          \"eng\",\n          \"jpn\",\n          \"ita\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"journal_title\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 4076,\n        \"samples\": [\n          \"Biopreservation and biobanking\",\n          \"Horticulture research\",\n          \"IET systems biology\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"pub_year\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 4.043198893232241,\n        \"min\": 1998.0,\n        \"max\": 2024.0,\n        \"num_unique_values\": 27,\n        \"samples\": [\n          2006.0,\n          2011.0,\n          2007.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"authors\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 79234,\n        \"samples\": [\n          \"Dona AC, Jim\\u00e9nez B, Sch\\u00e4fer H, Humpfer E, Spraul M, Lewis MR, Pearce JT, Holmes E, Lindon JC, Nicholson JK\",\n          \"Patel K, Lindsey J, Angelidou K, Aldrovandi G, Palumbo P\",\n          \"Meng X, Jayasundara N, Zhang J, Ren X, Gao B, Li J, Liu P\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"predicted_category\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 20,\n        \"samples\": [\n          \"Microbiology\",\n          \"Aging & Gerontology\",\n          \"Nephrology\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"full_embeddings\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"umap_2D_x\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 80385,\n        \"samples\": [\n          4.936419486999512,\n          4.239068984985352,\n          10.19593620300293\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"umap_2D_y\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 80482,\n        \"samples\": [\n          2.8839457035064697,\n          5.86353063583374,\n          8.207853317260742\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"umap_3D_x\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 80309,\n        \"samples\": [\n          5.784459590911865,\n          4.330934524536133,\n          7.4578094482421875\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"umap_3D_y\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 80424,\n        \"samples\": [\n          9.25174617767334,\n          5.252988815307617,\n          7.045469760894775\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"umap_3D_z\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 80350,\n        \"samples\": [\n          3.744140863418579,\n          3.4196255207061768,\n          4.59562349319458\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"tsne_2D_x\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 80533,\n        \"samples\": [\n          30.95949935913086,\n          -49.041751861572266,\n          -32.84935760498047\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"tsne_2D_y\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 80529,\n        \"samples\": [\n          62.61643600463867,\n          -50.186248779296875,\n          71.58711242675781\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}",
       "type": "dataframe",
       "variable_name": "df"
      },
      "text/html": [
       "\n",
       "  <div id=\"df-c98ad383-6b2f-4791-9217-16b9848ceaf1\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>language</th>\n",
       "      <th>journal_title</th>\n",
       "      <th>pub_year</th>\n",
       "      <th>authors</th>\n",
       "      <th>predicted_category</th>\n",
       "      <th>full_embeddings</th>\n",
       "      <th>umap_2D_x</th>\n",
       "      <th>umap_2D_y</th>\n",
       "      <th>umap_3D_x</th>\n",
       "      <th>umap_3D_y</th>\n",
       "      <th>umap_3D_z</th>\n",
       "      <th>tsne_2D_x</th>\n",
       "      <th>tsne_2D_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9748443</td>\n",
       "      <td>Effect of slow growth on metabolism of Escheri...</td>\n",
       "      <td>Escherichia coli growing on glucose in minimal...</td>\n",
       "      <td>eng</td>\n",
       "      <td>Journal of bacteriology</td>\n",
       "      <td>1998.0</td>\n",
       "      <td>Tweeddale H, Notley-McRobb L, Ferenci T</td>\n",
       "      <td>Microbiology</td>\n",
       "      <td>[[0.04921199, 0.1013429, 0.009529841, -0.08067...</td>\n",
       "      <td>7.770308</td>\n",
       "      <td>7.748135</td>\n",
       "      <td>8.087116</td>\n",
       "      <td>7.735348</td>\n",
       "      <td>6.041772</td>\n",
       "      <td>-18.861538</td>\n",
       "      <td>25.245789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10675895</td>\n",
       "      <td>On the optimization of classes for the assignm...</td>\n",
       "      <td>At present, the assignment of function to nove...</td>\n",
       "      <td>eng</td>\n",
       "      <td>Trends in biotechnology</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>Kell DB, King RD</td>\n",
       "      <td>unlabeled</td>\n",
       "      <td>[[0.074717656, 0.12005615, 0.023376802, 0.0167...</td>\n",
       "      <td>3.692910</td>\n",
       "      <td>6.724228</td>\n",
       "      <td>4.727096</td>\n",
       "      <td>6.715599</td>\n",
       "      <td>7.221296</td>\n",
       "      <td>45.593254</td>\n",
       "      <td>63.230408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10731098</td>\n",
       "      <td>Assessing the effect of reactive oxygen specie...</td>\n",
       "      <td>A two-dimensional thin-layer chromatographic a...</td>\n",
       "      <td>eng</td>\n",
       "      <td>Redox report : communications in free radical ...</td>\n",
       "      <td>1999.0</td>\n",
       "      <td>Tweeddale H, Notley-McRobb L, Ferenci T</td>\n",
       "      <td>unlabeled</td>\n",
       "      <td>[[-0.009071778, 0.013007838, -0.0069063944, -0...</td>\n",
       "      <td>8.485703</td>\n",
       "      <td>8.156181</td>\n",
       "      <td>8.033753</td>\n",
       "      <td>7.959665</td>\n",
       "      <td>5.779803</td>\n",
       "      <td>-18.892046</td>\n",
       "      <td>25.188398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-c98ad383-6b2f-4791-9217-16b9848ceaf1')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-c98ad383-6b2f-4791-9217-16b9848ceaf1 button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-c98ad383-6b2f-4791-9217-16b9848ceaf1');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "<div id=\"df-1849561f-2696-445a-acd6-bb8f94d6d12e\">\n",
       "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-1849561f-2696-445a-acd6-bb8f94d6d12e')\"\n",
       "            title=\"Suggest charts\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "     width=\"24px\">\n",
       "    <g>\n",
       "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
       "    </g>\n",
       "</svg>\n",
       "  </button>\n",
       "\n",
       "<style>\n",
       "  .colab-df-quickchart {\n",
       "      --bg-color: #E8F0FE;\n",
       "      --fill-color: #1967D2;\n",
       "      --hover-bg-color: #E2EBFA;\n",
       "      --hover-fill-color: #174EA6;\n",
       "      --disabled-fill-color: #AAA;\n",
       "      --disabled-bg-color: #DDD;\n",
       "  }\n",
       "\n",
       "  [theme=dark] .colab-df-quickchart {\n",
       "      --bg-color: #3B4455;\n",
       "      --fill-color: #D2E3FC;\n",
       "      --hover-bg-color: #434B5C;\n",
       "      --hover-fill-color: #FFFFFF;\n",
       "      --disabled-bg-color: #3B4455;\n",
       "      --disabled-fill-color: #666;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart {\n",
       "    background-color: var(--bg-color);\n",
       "    border: none;\n",
       "    border-radius: 50%;\n",
       "    cursor: pointer;\n",
       "    display: none;\n",
       "    fill: var(--fill-color);\n",
       "    height: 32px;\n",
       "    padding: 0;\n",
       "    width: 32px;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart:hover {\n",
       "    background-color: var(--hover-bg-color);\n",
       "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "    fill: var(--button-hover-fill-color);\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart-complete:disabled,\n",
       "  .colab-df-quickchart-complete:disabled:hover {\n",
       "    background-color: var(--disabled-bg-color);\n",
       "    fill: var(--disabled-fill-color);\n",
       "    box-shadow: none;\n",
       "  }\n",
       "\n",
       "  .colab-df-spinner {\n",
       "    border: 2px solid var(--fill-color);\n",
       "    border-color: transparent;\n",
       "    border-bottom-color: var(--fill-color);\n",
       "    animation:\n",
       "      spin 1s steps(1) infinite;\n",
       "  }\n",
       "\n",
       "  @keyframes spin {\n",
       "    0% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "      border-left-color: var(--fill-color);\n",
       "    }\n",
       "    20% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    30% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    40% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    60% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    80% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "    90% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "  }\n",
       "</style>\n",
       "\n",
       "  <script>\n",
       "    async function quickchart(key) {\n",
       "      const quickchartButtonEl =\n",
       "        document.querySelector('#' + key + ' button');\n",
       "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
       "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
       "      try {\n",
       "        const charts = await google.colab.kernel.invokeFunction(\n",
       "            'suggestCharts', [key], {});\n",
       "      } catch (error) {\n",
       "        console.error('Error during call to suggestCharts:', error);\n",
       "      }\n",
       "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
       "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
       "    }\n",
       "    (() => {\n",
       "      let quickchartButtonEl =\n",
       "        document.querySelector('#df-1849561f-2696-445a-acd6-bb8f94d6d12e button');\n",
       "      quickchartButtonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "    })();\n",
       "  </script>\n",
       "</div>\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ],
      "text/plain": [
       "       pmid                                              title  \\\n",
       "0   9748443  Effect of slow growth on metabolism of Escheri...   \n",
       "1  10675895  On the optimization of classes for the assignm...   \n",
       "2  10731098  Assessing the effect of reactive oxygen specie...   \n",
       "\n",
       "                                            abstract language  \\\n",
       "0  Escherichia coli growing on glucose in minimal...      eng   \n",
       "1  At present, the assignment of function to nove...      eng   \n",
       "2  A two-dimensional thin-layer chromatographic a...      eng   \n",
       "\n",
       "                                       journal_title  pub_year  \\\n",
       "0                            Journal of bacteriology    1998.0   \n",
       "1                            Trends in biotechnology    2000.0   \n",
       "2  Redox report : communications in free radical ...    1999.0   \n",
       "\n",
       "                                   authors predicted_category  \\\n",
       "0  Tweeddale H, Notley-McRobb L, Ferenci T       Microbiology   \n",
       "1                         Kell DB, King RD          unlabeled   \n",
       "2  Tweeddale H, Notley-McRobb L, Ferenci T          unlabeled   \n",
       "\n",
       "                                     full_embeddings  umap_2D_x  umap_2D_y  \\\n",
       "0  [[0.04921199, 0.1013429, 0.009529841, -0.08067...   7.770308   7.748135   \n",
       "1  [[0.074717656, 0.12005615, 0.023376802, 0.0167...   3.692910   6.724228   \n",
       "2  [[-0.009071778, 0.013007838, -0.0069063944, -0...   8.485703   8.156181   \n",
       "\n",
       "   umap_3D_x  umap_3D_y  umap_3D_z  tsne_2D_x  tsne_2D_y  \n",
       "0   8.087116   7.735348   6.041772 -18.861538  25.245789  \n",
       "1   4.727096   6.715599   7.221296  45.593254  63.230408  \n",
       "2   8.033753   7.959665   5.779803 -18.892046  25.188398  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "7S8D2-cnPMUK"
   },
   "outputs": [],
   "source": [
    "df_filtered = df[['title', 'abstract', 'journal_title', 'pub_year', 'authors', 'tsne_2D_x', 'tsne_2D_y']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 372
    },
    "executionInfo": {
     "elapsed": 1074,
     "status": "ok",
     "timestamp": 1728493851711,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "bb139znXPaDj",
    "outputId": "922f070f-f2ce-4b17-d383-a91aeabefffb"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "summary": "{\n  \"name\": \"df_filtered\",\n  \"rows\": 80656,\n  \"fields\": [\n    {\n      \"column\": \"title\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 80599,\n        \"samples\": [\n          \"Responsive MRI agents for sensing metabolism in vivo.\",\n          \"Metabolomics and Machine Learning Identify Metabolic Differences and Potential Biomarkers for Frequent Versus Infrequent Gout Flares.\",\n          \"Postprandial NMR-Based Metabolic Exchanges Reflect Impaired Phenotypic Flexibility across Splanchnic Organs in the Obese Yucatan Mini-Pig.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"abstract\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 80583,\n        \"samples\": [\n          \"Drought is a misfortune for agriculture and human beings. The annual crop yield reduction caused by drought exceeds the sum of all pathogens. As one of the gatekeepers of China's \\\"granary\\\", rice is the most important to reveal the key drought tolerance factors in rice. Rice seedlings of Nipponbare (Oryza sativa L. ssp. Japonica) were subjected to simulated drought stress, and their root systems were analyzed for the non-targeted metabolome and strand-specific transcriptome. We found that both DEGs and metabolites were enriched in purine metabolism, and allantoin accumulated significantly in roots under drought stress. However, few studies on drought tolerance of exogenous allantoin in rice have been reported. We aimed to further determine whether allantoin can improve the drought tolerance of rice. Under the treatment of exogenous allantoin at different concentrations, the drought resistant metabolites of plants accumulated significantly, including proline and soluble sugar, and reactive oxygen species (ROS) decreased and reached a significant level in 100 \\u03bcmol L[-1]. To this end, a follow-up study was identified in 100 \\u03bcmol L[-1] exogenous allantoin and found that exogenous allantoin improved the drought resistance of rice. At the gene level, under allantoin drought treatment, we found that genes of scavenge reactive oxygen species were significantly expressed, including peroxidase (POD), catalase (CATA), ascorbate peroxidase 8 (APX8) and respiratory burst oxidase homolog protein F (RbohF). This indicates that plants treated by allantoin have better ability to scavenge reactive oxygen species to resist drought. Alternative splicing analysis revealed a total of 427 differentially expressed alternative splicing events across 320 genes. The analysis of splicing factors showed that gene alternative splicing could be divided into many different subgroups and play a regulatory role in many aspects. Through further analysis, we restated the key genes and enzymes in the allantoin synthesis and catabolism pathway, and found that the expression of synthetase and hydrolase showed a downward trend. The pathway of uric acid to allantoin is completed by uric acid oxidase (UOX). To find out the key transcription factors that regulate the expression of this gene, we identified two highly related transcription factors OsERF059 and ONAC007 through correlation analysis. They may be the key for allantoin to enhance the drought resistance of rice.\",\n          \"The study aim was to evaluate the effects of public lockdown during the covid-19 pandemic on glucose and metabolic parameters as well as body weight control in type 2 diabetic patients.\",\n          \"1. Geniposide (genipin 1-O-glucose), one of the major bioactive constituents isolated from Fructus Gardeniae, possesses many biological activities. In this study, an efficient strategy was developed using ultra-high-performance liquid chromatography coupled with linear ion trap-Orbitrap mass spectrometer (UPLC-LTQ-Orbitrap) to profile the in vitro and in vivo metabolic patterns of geniposide in rat liver microsomes (RLMs), plasma, urine, and various tissues. And post-acquisition data-mining methods including extracted ion chromatogram (EIC), multiple mass defect filters (MMDF), fragment ion searching (FISh), and isotope pattern filtering (IPF) were adopted to characterize the known and unknown metabolites. 2. A total of 33 metabolites were detected and interpreted according to accurate mass measurement, diagnostic fragment ions, relevant drug biotransformation knowledge, and bibliography data. Among them, 17 metabolites were detected in the plasma, 31 metabolites were identified in the urine, six metabolites could be found in rat heart, 12 in liver, three in spleen, six in lung, 12 in kidney, six in brain, and four in RLMs. 3. A series of corresponding reactions such as hydrolysis, hydroxylation, taurine conjugation, hydrogenation, decarboxylation, demethylation, sulfate conjugation, cysteine S-conjugation, glucosylation, and their composite reactions were all discovered. 4. The results could provide comprehensive insights and guidance for elucidation of side effect mechanism and safety monitoring as well as for rational formulation design in drug delivery system. The newly discovered geniposide metabolites could be targets for future metabolism studies on the important chemical constituents from herbal medicines.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"journal_title\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 4076,\n        \"samples\": [\n          \"Biopreservation and biobanking\",\n          \"Horticulture research\",\n          \"IET systems biology\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"pub_year\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 4.043198893232241,\n        \"min\": 1998.0,\n        \"max\": 2024.0,\n        \"num_unique_values\": 27,\n        \"samples\": [\n          2006.0,\n          2011.0,\n          2007.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"authors\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 79234,\n        \"samples\": [\n          \"Dona AC, Jim\\u00e9nez B, Sch\\u00e4fer H, Humpfer E, Spraul M, Lewis MR, Pearce JT, Holmes E, Lindon JC, Nicholson JK\",\n          \"Patel K, Lindsey J, Angelidou K, Aldrovandi G, Palumbo P\",\n          \"Meng X, Jayasundara N, Zhang J, Ren X, Gao B, Li J, Liu P\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"tsne_2D_x\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 80533,\n        \"samples\": [\n          30.95949935913086,\n          -49.041751861572266,\n          -32.84935760498047\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"tsne_2D_y\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 80529,\n        \"samples\": [\n          62.61643600463867,\n          -50.186248779296875,\n          71.58711242675781\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}",
       "type": "dataframe",
       "variable_name": "df_filtered"
      },
      "text/html": [
       "\n",
       "  <div id=\"df-1f0c3696-00e4-4f20-a4c0-274ba4464500\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>journal_title</th>\n",
       "      <th>pub_year</th>\n",
       "      <th>authors</th>\n",
       "      <th>tsne_2D_x</th>\n",
       "      <th>tsne_2D_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Effect of slow growth on metabolism of Escheri...</td>\n",
       "      <td>Escherichia coli growing on glucose in minimal...</td>\n",
       "      <td>Journal of bacteriology</td>\n",
       "      <td>1998.0</td>\n",
       "      <td>Tweeddale H, Notley-McRobb L, Ferenci T</td>\n",
       "      <td>-18.861538</td>\n",
       "      <td>25.245789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>On the optimization of classes for the assignm...</td>\n",
       "      <td>At present, the assignment of function to nove...</td>\n",
       "      <td>Trends in biotechnology</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>Kell DB, King RD</td>\n",
       "      <td>45.593254</td>\n",
       "      <td>63.230408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Assessing the effect of reactive oxygen specie...</td>\n",
       "      <td>A two-dimensional thin-layer chromatographic a...</td>\n",
       "      <td>Redox report : communications in free radical ...</td>\n",
       "      <td>1999.0</td>\n",
       "      <td>Tweeddale H, Notley-McRobb L, Ferenci T</td>\n",
       "      <td>-18.892046</td>\n",
       "      <td>25.188398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-1f0c3696-00e4-4f20-a4c0-274ba4464500')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-1f0c3696-00e4-4f20-a4c0-274ba4464500 button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-1f0c3696-00e4-4f20-a4c0-274ba4464500');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "<div id=\"df-82cab806-f423-44ca-9897-a8f980ab3883\">\n",
       "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-82cab806-f423-44ca-9897-a8f980ab3883')\"\n",
       "            title=\"Suggest charts\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "     width=\"24px\">\n",
       "    <g>\n",
       "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
       "    </g>\n",
       "</svg>\n",
       "  </button>\n",
       "\n",
       "<style>\n",
       "  .colab-df-quickchart {\n",
       "      --bg-color: #E8F0FE;\n",
       "      --fill-color: #1967D2;\n",
       "      --hover-bg-color: #E2EBFA;\n",
       "      --hover-fill-color: #174EA6;\n",
       "      --disabled-fill-color: #AAA;\n",
       "      --disabled-bg-color: #DDD;\n",
       "  }\n",
       "\n",
       "  [theme=dark] .colab-df-quickchart {\n",
       "      --bg-color: #3B4455;\n",
       "      --fill-color: #D2E3FC;\n",
       "      --hover-bg-color: #434B5C;\n",
       "      --hover-fill-color: #FFFFFF;\n",
       "      --disabled-bg-color: #3B4455;\n",
       "      --disabled-fill-color: #666;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart {\n",
       "    background-color: var(--bg-color);\n",
       "    border: none;\n",
       "    border-radius: 50%;\n",
       "    cursor: pointer;\n",
       "    display: none;\n",
       "    fill: var(--fill-color);\n",
       "    height: 32px;\n",
       "    padding: 0;\n",
       "    width: 32px;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart:hover {\n",
       "    background-color: var(--hover-bg-color);\n",
       "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "    fill: var(--button-hover-fill-color);\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart-complete:disabled,\n",
       "  .colab-df-quickchart-complete:disabled:hover {\n",
       "    background-color: var(--disabled-bg-color);\n",
       "    fill: var(--disabled-fill-color);\n",
       "    box-shadow: none;\n",
       "  }\n",
       "\n",
       "  .colab-df-spinner {\n",
       "    border: 2px solid var(--fill-color);\n",
       "    border-color: transparent;\n",
       "    border-bottom-color: var(--fill-color);\n",
       "    animation:\n",
       "      spin 1s steps(1) infinite;\n",
       "  }\n",
       "\n",
       "  @keyframes spin {\n",
       "    0% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "      border-left-color: var(--fill-color);\n",
       "    }\n",
       "    20% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    30% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    40% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    60% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    80% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "    90% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "  }\n",
       "</style>\n",
       "\n",
       "  <script>\n",
       "    async function quickchart(key) {\n",
       "      const quickchartButtonEl =\n",
       "        document.querySelector('#' + key + ' button');\n",
       "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
       "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
       "      try {\n",
       "        const charts = await google.colab.kernel.invokeFunction(\n",
       "            'suggestCharts', [key], {});\n",
       "      } catch (error) {\n",
       "        console.error('Error during call to suggestCharts:', error);\n",
       "      }\n",
       "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
       "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
       "    }\n",
       "    (() => {\n",
       "      let quickchartButtonEl =\n",
       "        document.querySelector('#df-82cab806-f423-44ca-9897-a8f980ab3883 button');\n",
       "      quickchartButtonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "    })();\n",
       "  </script>\n",
       "</div>\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ],
      "text/plain": [
       "                                               title  \\\n",
       "0  Effect of slow growth on metabolism of Escheri...   \n",
       "1  On the optimization of classes for the assignm...   \n",
       "2  Assessing the effect of reactive oxygen specie...   \n",
       "\n",
       "                                            abstract  \\\n",
       "0  Escherichia coli growing on glucose in minimal...   \n",
       "1  At present, the assignment of function to nove...   \n",
       "2  A two-dimensional thin-layer chromatographic a...   \n",
       "\n",
       "                                       journal_title  pub_year  \\\n",
       "0                            Journal of bacteriology    1998.0   \n",
       "1                            Trends in biotechnology    2000.0   \n",
       "2  Redox report : communications in free radical ...    1999.0   \n",
       "\n",
       "                                   authors  tsne_2D_x  tsne_2D_y  \n",
       "0  Tweeddale H, Notley-McRobb L, Ferenci T -18.861538  25.245789  \n",
       "1                         Kell DB, King RD  45.593254  63.230408  \n",
       "2  Tweeddale H, Notley-McRobb L, Ferenci T -18.892046  25.188398  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_filtered.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "VeXCG1VmWqdV"
   },
   "outputs": [],
   "source": [
    "abstracts = df_filtered[\"abstract\"]\n",
    "titles = df_filtered[\"title\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 7,
     "status": "ok",
     "timestamp": 1728493851712,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "WFQAVNq6XFXJ",
    "outputId": "8b80551a-cc7b-443d-ae22-206289e1427a"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Escherichia coli growing on glucose in minimal medium controls its metabolite pools in response to environmental conditions. The extent of pool changes was followed through two-dimensional thin-layer chromatography of all 14C-glucose labelled compounds extracted from bacteria. The patterns of metabolites and spot intensities detected by phosphorimaging were found to reproducibly differ depending on culture conditions. Clear trends were apparent in the pool sizes of several of the 70 most abundant metabolites extracted from bacteria growing in glucose-limited chemostats at different growth rates. The pools of glutamate, aspartate, trehalose, and adenosine as well as UDP-sugars and putrescine changed markedly. The data on pools observed by two-dimensional thin-layer chromatography were confirmed for amino acids by independent analysis. Other unidentified metabolites also displayed different spot intensities under various conditions, with four trend patterns depending on growth rate. As RpoS controls a number of metabolic genes in response to nutrient limitation, an rpoS mutant was also analyzed for metabolite pools. The mutant had altered metabolite profiles, but only some of the changes at slow growth rates were ascribable to the known control of metabolic genes by RpoS. These results indicate that total metabolite pool (\"metabolome\") analysis offers a means of revealing novel aspects of cellular metabolism and global regulation.\n"
     ]
    }
   ],
   "source": [
    "print(abstracts[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 6,
     "status": "ok",
     "timestamp": 1728493851712,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "ZWj-9tiDXRxe",
    "outputId": "42fc0ec5-9024-4ae7-a2a1-fb0df4e50c2c"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "80656"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(abstracts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "aLQjuKzosKQu",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## OpenAI GPT - setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "SsXMWAtBXmNF"
   },
   "outputs": [],
   "source": [
    "from google.colab import userdata\n",
    "import os\n",
    "\n",
    "openai_key = userdata.get('OPENAI_API_KEY')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "rZKpvjTHrY6e"
   },
   "source": [
    "### Prompt Template: OpenAI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "bQRaZzLVTtYw"
   },
   "outputs": [],
   "source": [
    "# OpenAI prompt\n",
    "\n",
    "openai_prompt = \"\"\"\n",
    "You are an expert in metabolomics and scientific literature analysis. Your task\n",
    "is to generate concise, informative topic labels for collections of metabolomics\n",
    "abstracts from PubMed. Each topic label should be no more than 6 words long and\n",
    "should capture the essence of the metabolomics research described.\n",
    "\n",
    "Here is an example:\n",
    "I have a topic that contains the following metabolomics abstracts:\n",
    "- This study investigates the metabolic profiling of plasma samples from patients\n",
    "with type 2 diabetes using LC-MS/MS. We identified several key metabolites\n",
    "associated with insulin resistance.\n",
    "- Our research focuses on the application of NMR spectroscopy to analyze urine\n",
    "samples for early detection of kidney disease. The metabolic signatures\n",
    "revealed potential biomarkers.\n",
    "- We employed GC-MS to examine the metabolome of cancer cells under hypoxic\n",
    "conditions. The results showed significant alterations in glucose and glutamine\n",
    "metabolism.\n",
    "\n",
    "The topic is described by the following keywords: 'metabolomics, LC-MS, NMR,\n",
    "biomarkers, disease detection.'\n",
    "\n",
    "A suitable topic label would be: Disease Biomarker Discovery.\n",
    "\n",
    "Now, based on the information provided below, please create a concise topic\n",
    "label for this metabolomics topic in 6 words or fewer.\n",
    "\n",
    "Documents: [DOCUMENTS]\n",
    "Keywords: [KEYWORDS]\n",
    "\n",
    "Return only the topic label, nothing more. Make sure it is in the following format:\n",
    "topic: <topic label>\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "6LKyccJidu3K"
   },
   "source": [
    "## 🗨️ BERTopic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ERBG_XcZeEyE"
   },
   "source": [
    "### Sub-models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jl-WT_1Gd466",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### PubMed Embeddings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "5-MDirESLIYp"
   },
   "source": [
    "The reasoning behind `PrecomputedEmbeddings` class.\n",
    "\n",
    "\n",
    "\n",
    "The PrecomputedEmbeddings class was introduced as a workaround to a specific challenge I faced when trying to use BERTopic with our pre-computed embeddings.\n",
    "\n",
    "1. BERTopic's Default Behavior:\n",
    "   By default, BERTopic is designed to generate embeddings for documents as part of its pipeline. It typically uses a pre-trained language model (like BERT or Sentence-BERT) to create these embeddings during the fit_transform process.\n",
    "\n",
    "2. Our Scenario:\n",
    "   In our case, I already have pre-computed embeddings for our abstracts using pubMEDBERT.\n",
    "\n",
    "3. The Issue:\n",
    "   When I tried to pass the pre-computed embeddings directly to BERTopic, I encountered errors. This is because BERTopic was still trying to generate new embeddings using its default embedding model, which was set to `None` in our case.\n",
    "\n",
    "4. The Solution:\n",
    "   The PrecomputedEmbeddings class acts as a \"fake\" embedding model for BERTopic. It mimics the interface of a typical embedding model but instead of generating new embeddings, it simply returns the pre-computed embeddings we provide.\n",
    "\n",
    "Here's how the class works:\n",
    "\n",
    "```python\n",
    "class PrecomputedEmbeddings:\n",
    "    def __init__(self, embeddings):\n",
    "        self.embeddings = embeddings\n",
    "\n",
    "    def embed_documents(self, documents, verbose=False):\n",
    "        return self.embeddings\n",
    "\n",
    "    def embed_words(self, words, verbose=False):\n",
    "        return self.embeddings\n",
    "```\n",
    "\n",
    "- The `__init__` method stores the pre-computed embeddings.\n",
    "- The `embed_documents` and `embed_words` methods are required by BERTopic's interface, but they simply return the stored embeddings regardless of the input.\n",
    "\n",
    "By using this class, we're essentially telling BERTopic: \"Don't generate new embeddings. Use these pre-computed ones instead.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "AyE8raUCAeWl"
   },
   "outputs": [],
   "source": [
    "class PrecomputedEmbeddings:\n",
    "    def __init__(self, embeddings):\n",
    "        self.embeddings = embeddings\n",
    "\n",
    "    def embed_documents(self, documents, verbose=False):\n",
    "        return self.embeddings\n",
    "\n",
    "    def embed_words(self, words, verbose=False):\n",
    "        return self.embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "IKZczLiEc3sw"
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "embeddings = np.array(df['full_embeddings'].tolist())\n",
    "if embeddings.ndim == 3:\n",
    "    embeddings = embeddings.reshape(embeddings.shape[0], -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 9,
     "status": "ok",
     "timestamp": 1728494069478,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "4-yfMEkA7vRY",
    "outputId": "2e1ee043-9dab-4d35-b5f3-9f8fb7255152"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1, 768)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['full_embeddings'][0].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 6,
     "status": "ok",
     "timestamp": 1728494069478,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "TBrTPK26-zdJ",
    "outputId": "be0f96c2-0c34-419b-d601-970c8212809b"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(80656, 768)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "embeddings.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "GW--svE6AgpV"
   },
   "outputs": [],
   "source": [
    "# Create an instance of PrecomputedEmbeddings\n",
    "embedding_model = PrecomputedEmbeddings(embeddings)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "-SzWRahc-NnP",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Dimensionality Reduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "S8opbjCEeQho"
   },
   "source": [
    "**Pre-computed reduced embeddings**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "OYXx76OzeB8q"
   },
   "outputs": [],
   "source": [
    "#reduced embeddings\n",
    "reduced_embeddings = df[['tsne_2D_x', 'tsne_2D_y']].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "_lJ-wcd47x5W",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Clustering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MryO3U7E8mbR"
   },
   "source": [
    "**The resoning behind clustering**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "AJ9u5EK6-XLs"
   },
   "source": [
    "**HDBSCAN vs K-Means:** HDBSCAN was selected due to its strong ability to capture structures with varying densities, making it particularly useful for this context. It’s important to note that no clustering model is perfect. For example, K-means allows you to predefine the number of clusters and forces every point into a cluster, meaning no outliers are created. However, this method has its drawbacks. By forcing every point into a cluster, the model is likely to include noise, which can distort topic representations and negatively impact the quality of the clustering.\n",
    "\n",
    "**Starting Point**: Given the dataset size of ~80k and the desire for fewer topics, a good starting point would be to set `min_cluster_size` to around 800 (1% of your dataset).\n",
    "\n",
    "**Iterative Approach**: Start with this conservative estimate and then adjust based on the results you get:\n",
    "\n",
    "- If you get too many topics, increase `min_cluster_size`\n",
    "- If you get too few topics, decrease `min_cluster_size`\n",
    "\n",
    "Metabolomics is a broad field, so we might want to lean towards larger clusters (higher `min_cluster_size`) to capture overarching themes.\n",
    "\n",
    "**Additional Parameters**: Consider setting `min_samples` equal to `min_cluster_size` for more conservative clustering. You can also experiment with c`luster_selection_epsilon` to merge smaller clusters into larger ones."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "J0oG8yRTBySH"
   },
   "source": [
    "**Understanding `min_samples` and its relationship with `min_cluster_size`**\n",
    "\n",
    "**Purpose**: `min_samples` determines how many neighboring points a data point needs to be considered a core point in the clustering process. <br>\n",
    "**Effect on Clustering**: A higher `min_samples` value leads to more conservative clustering, with fewer but more robust clusters. <br>\n",
    "**Setting it Equal to `min_cluster_size`**:\n",
    "- This ensures that every point in a cluster is a core point.\n",
    "- It results in more clearly defined, robust clusters.\n",
    "- It reduces the risk of chaining (where disparate clusters get artificially connected).\n",
    "- It simplifies the parameter tuning process. <br>\n",
    "\n",
    "**Trade-offs**: While this approach leads to more coherent clusters, it may also result in fewer clusters overall and more points being labeled as noise. <br>\n",
    "\n",
    "**Flexibility**: You can adjust `min_samples` independently of `min_cluster_size` if needed:\n",
    "- Lowering it allows for more variation within clusters.\n",
    "- Raising it makes clustering even more conservative.\n",
    "\n",
    "\n",
    "\n",
    "For this specific case with ~80,000 metabolomics abstracts, starting with `min_samples` = `min_cluster_size` = 800 is a good conservative approach. This will likely give fewer, but very well-defined topics. If this results in too few clusters or too many abstracts being labeled as noise, you can try slightly decreasing `min_samples` while keeping `min_cluster_size` the same.\n",
    "Remember, the best settings often come through an iterative process of adjusting parameters and evaluating results. The goal is to find clusters (topics) that are meaningful and interpretable in the context of metabolomics research."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "IuyI97nreCAI"
   },
   "outputs": [],
   "source": [
    "#from umap import UMAP\n",
    "from hdbscan import HDBSCAN\n",
    "\n",
    "hdbscan_model = HDBSCAN(min_cluster_size=500, #change this parameter to update number of topics.\n",
    "                        min_samples=300,\n",
    "                        metric='euclidean',\n",
    "                        cluster_selection_method='eom',\n",
    "                        prediction_data=True) #prediction_data=True ensures that the model can generalize to new documents, assigning them to the appropriate topics as needed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "RVyAstZtf3TP",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Tokenizer/ Vectorizer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "LiOjRdb1gYn2"
   },
   "source": [
    "**On Count Vectorizer Model**\n",
    "\n",
    "Parameters:\n",
    "\n",
    "**stop_words=\"english\"**: This removes common English stop words (like \"the\", \"a\", \"an\", \"in\") from the text. Stop words often don't contribute much to the meaning of a topic, so removing them helps focus on more meaningful words.\n",
    "\n",
    "**min_df=_n_**: This sets the minimum document frequency for a term to be included. A term must appear in at least _n_ documents to be considered. This helps remove very rare words that might not be representative of broader topics.\n",
    "\n",
    "**ngram_range=(1, 2)**: This allows the vectorizer to consider both unigrams (single words) and bigrams (two-word phrases). This can capture more complex concepts that might be expressed in phrases rather than single words.\n",
    "\n",
    "How this improves topic modeling:\n",
    "\n",
    "- Removing stop words helps focus on more meaningful content words, making topics more interpretable and distinct.\n",
    "- Setting a minimum document frequency (min_df) helps filter out very rare terms that might be noise or very specific to a single document, leading to more general and robust topics.\n",
    "- Including bigrams allows the model to capture more complex concepts and phrases, which can lead to more nuanced and interpretable topics. For example, instead of just \"learning\" and \"machine\" as separate words, it might capture \"machine learning\" as a single concept.\n",
    "- By preprocessing the text in this way, the c-TF-IDF step in BERTopic (which uses **CountVectorizer**) can create more meaningful and distinct topic representations.\n",
    "This preprocessing happens after documents are assigned to topics, so it doesn't influence the clustering process itself. Instead, it improves how topics are represented and described once they're formed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "157KK6flX4e3"
   },
   "source": [
    "_Note that I can update the vectorizer model after fitting the BERTopic model. This will allow for better topic representation without retraining the model._"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "QsHdqgZw0aKd"
   },
   "outputs": [],
   "source": [
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "\n",
    "# Create CountVectorizer\n",
    "vectorizer_model = CountVectorizer(\n",
    "    stop_words=\"english\",  # This removes common English stop words\n",
    "    min_df=10,             # A term must appear in at least 10 documents to be considered\n",
    "    ngram_range=(1, 2)     # Consider both unigrams (single words) and bigrams (two-word phrases)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "9gZfgW8o4-tl",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Topic Representation: c-TF-IDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "XJO3-n-2mes0"
   },
   "source": [
    "for more information is [c-TF-IDF BERTopic documentation](https://maartengr.github.io/BERTopic/getting_started/ctfidf/ctfidf.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "cEgiRvyA5Ehe"
   },
   "outputs": [],
   "source": [
    "from bertopic.vectorizers import ClassTfidfTransformer\n",
    "\n",
    "ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "vWemkaDIepqm",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Distilled Representation Model & Fine Tuning"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "pmDUyN46a7F8"
   },
   "source": [
    "Running 50 abstracts through GPT-4o costs approximately $0.90, and processing 100 abstracts is roughly double that amount.\n",
    "\n",
    "I experimented with 4, 50, and 100 abstracts, and found no significant differences in topic generation, particularly when comparing 50 versus 100 abstracts\n",
    "\n",
    "On the other hand running 100 abstracts with GPT-4o mini costs just about 10 cents!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "1sxYYx4cdPe6"
   },
   "outputs": [],
   "source": [
    "from bertopic.representation import KeyBERTInspired, TextGeneration, OpenAI\n",
    "import openai\n",
    "\n",
    "# KeyBERT\n",
    "#keybert = KeyBERTInspired()\n",
    "\n",
    "#GPT-4\n",
    "client = openai.OpenAI(api_key=openai_key)\n",
    "gpt4o = OpenAI(client,\n",
    "               model=\"gpt-4o-mini\", #Use gpt-4o\n",
    "               delay_in_seconds=2,\n",
    "               exponential_backoff=True, #retries protocol\n",
    "               chat=True,\n",
    "               prompt=openai_prompt,\n",
    "               nr_docs=100 #number of abstracts included in the [DOCUMENTS]\n",
    "               )\n",
    "\n",
    "# Text generation with Llama 3\n",
    "#llama3_1 = TextGeneration(generator,\n",
    "#                          prompt=prompt,\n",
    "#                          nr_docs=5 #number of abstracts included in the [DOCUMENTS]\n",
    "#                          )\n",
    "\n",
    "# All representation models\n",
    "representation_model = {\n",
    "    #\"KeyBERT\": keybert,\n",
    "    \"GPT-4o\": gpt4o,\n",
    "    #\"Llama3.1\": llama3_1,\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "fqkWKW7CfF9M",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 🔥 Training"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "xsOtpNuP3MgO"
   },
   "source": [
    "**How the model comes up with the topics**:\n",
    "\n",
    "It doesn't process all abstracts directly in this case. Instead, it works on a per-topic basis, where topics have been pre-determined by the clustering algorithm (HDBSCAN in this case).\n",
    "For each topic, Llama receives:\n",
    "- A set of keywords that represent the topic (generated by c-TF-IDF).\n",
    "- A small subset of representative documents from that topic (~4-5??).\n",
    "\n",
    "The model then generates a concise label (5 words or less) based on this information.\n",
    "This process is repeated for each topic identified by the clustering algorithm.\n",
    "\n",
    "The advantage of this approach is that it allows for efficient topic labeling even with large datasets. Llama doesn't need to process all documents, but instead works with a *distilled representation* of each topic.\n",
    "However, this also means that the quality of Llama's labels depends heavily on the preceding steps: the quality of the initial embeddings, the effectiveness of the clustering, and the *representativeness of the selected documents and keywords for each topic*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 212,
     "referenced_widgets": [
      "e08b89598596417ea2183100097f33be",
      "8e520224d0324e139a9b4528639e4d75",
      "84754e2366ab4bc1b9f1b0c6470f4879",
      "ca327b5c507b4dc28b7400ea0df30b66",
      "ef49e2bf2d44470da616ae07ddb270a9",
      "3c68485af6724380ac269f1747e4b479",
      "d94785a14eec4a6ab914f1691b4cc811",
      "ffb91fa10ae6449fbfb89113ee16062d",
      "175db7c0169646cc8c340cb10df6b68f",
      "95fd75e407bd45ba9803b8d347f99173",
      "169fed61ddbb4f96a9a178eda9d3af5d"
     ]
    },
    "executionInfo": {
     "elapsed": 475027,
     "status": "ok",
     "timestamp": 1728498605227,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "rZNPILMDY_u4",
    "outputId": "5ed0c534-efca-46ce-a6f8-7c40595fa155"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-09 18:22:10,122 - BERTopic - Embedding - Transforming documents to embeddings.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e08b89598596417ea2183100097f33be",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Batches:   0%|          | 0/2521 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-09 18:25:39,288 - BERTopic - Embedding - Completed ✓\n",
      "2024-10-09 18:25:39,290 - BERTopic - Dimensionality - Fitting the dimensionality reduction algorithm\n",
      "2024-10-09 18:26:23,983 - BERTopic - Dimensionality - Completed ✓\n",
      "2024-10-09 18:26:23,987 - BERTopic - Cluster - Start clustering the reduced embeddings\n",
      "2024-10-09 18:26:37,626 - BERTopic - Cluster - Completed ✓\n",
      "2024-10-09 18:26:37,650 - BERTopic - Representation - Extracting topics from clusters using representation models.\n",
      "100%|██████████| 21/21 [02:55<00:00,  8.34s/it]\n",
      "2024-10-09 18:30:02,577 - BERTopic - Representation - Completed ✓\n"
     ]
    }
   ],
   "source": [
    "from bertopic import BERTopic\n",
    "\n",
    "# Create and train BERTopic model\n",
    "topic_model = BERTopic(\n",
    "    embedding_model=embedding_model, # Step 1 - Extract embeddings\n",
    "    umap_model=None,                 # Step 2 - Reduce dimensionality\n",
    "    hdbscan_model=hdbscan_model,     # Step 3 - Cluster documents\n",
    "    vectorizer_model=vectorizer_model, # Step 4 - Tokenize topics\n",
    "    ctfidf_model=ctfidf_model,         # Step 5 - Extract topic words\n",
    "    representation_model=representation_model, # Step 6 - Fine-tune topic representations\n",
    "\n",
    "    # Hyperparameters\n",
    "    top_n_words=10,\n",
    "    verbose=True\n",
    ")\n",
    "\n",
    "# Train model using pre-computed embeddings\n",
    "topics, probs = topic_model.fit_transform(abstracts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VuXOqG3r1Q77"
   },
   "source": [
    "## ☑ Results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "BpOmV4Dmjtt9",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Topics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 293
    },
    "executionInfo": {
     "elapsed": 346,
     "status": "ok",
     "timestamp": 1728498617546,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "rhWPK0OvfFJV",
    "outputId": "904e9077-cb44-4f80-b289-5110339b7152"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "summary": "{\n  \"name\": \"topic_model\",\n  \"rows\": 5,\n  \"fields\": [\n    {\n      \"column\": \"Topic\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 1,\n        \"min\": -1,\n        \"max\": 3,\n        \"num_unique_values\": 5,\n        \"samples\": [\n          0,\n          3,\n          1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Count\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 5441,\n        \"min\": 6499,\n        \"max\": 19258,\n        \"num_unique_values\": 5,\n        \"samples\": [\n          19258,\n          6499,\n          9730\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Name\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"0_plant_plants_genes_compounds\",\n          \"3_data_ms_mass_metabolomics\",\n          \"1_liver_diabetes_insulin_muscle\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Representation\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"GPT-4o\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Representative_Docs\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}",
       "type": "dataframe"
      },
      "text/html": [
       "\n",
       "  <div id=\"df-e01f6712-beb4-4399-b493-998ddee933a5\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Count</th>\n",
       "      <th>Name</th>\n",
       "      <th>Representation</th>\n",
       "      <th>GPT-4o</th>\n",
       "      <th>Representative_Docs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "      <td>14980</td>\n",
       "      <td>-1_exposure_metabolic_study_patients</td>\n",
       "      <td>[exposure, metabolic, study, patients, metabol...</td>\n",
       "      <td>[Metabolomics in Disease and Treatment]</td>\n",
       "      <td>[Background: In numerous studies based predomi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>19258</td>\n",
       "      <td>0_plant_plants_genes_compounds</td>\n",
       "      <td>[plant, plants, genes, compounds, stress, grow...</td>\n",
       "      <td>[Plant Stress Response Mechanisms]</td>\n",
       "      <td>[To investigate differences in fresh leaves of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>9730</td>\n",
       "      <td>1_liver_diabetes_insulin_muscle</td>\n",
       "      <td>[liver, diabetes, insulin, muscle, mice, exerc...</td>\n",
       "      <td>[Metabolic Profiles and Dysregulation]</td>\n",
       "      <td>[Fenugreek is a well-known medicinal plant use...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>7254</td>\n",
       "      <td>2_cancer_tumor_cells_cell</td>\n",
       "      <td>[cancer, tumor, cells, cell, breast, cancer ce...</td>\n",
       "      <td>[Cancer Metabolism and Therapy Resistance]</td>\n",
       "      <td>[Cancer cells rewire the metabolic processes b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>6499</td>\n",
       "      <td>3_data_ms_mass_metabolomics</td>\n",
       "      <td>[data, ms, mass, metabolomics, sample, method,...</td>\n",
       "      <td>[Metabolomics Data Analysis and Integration]</td>\n",
       "      <td>[Although metabolomics data acquisition and an...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-e01f6712-beb4-4399-b493-998ddee933a5')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-e01f6712-beb4-4399-b493-998ddee933a5 button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-e01f6712-beb4-4399-b493-998ddee933a5');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "<div id=\"df-1a3bfaba-0225-4b93-aea9-0a0d48f10bc2\">\n",
       "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-1a3bfaba-0225-4b93-aea9-0a0d48f10bc2')\"\n",
       "            title=\"Suggest charts\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "     width=\"24px\">\n",
       "    <g>\n",
       "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
       "    </g>\n",
       "</svg>\n",
       "  </button>\n",
       "\n",
       "<style>\n",
       "  .colab-df-quickchart {\n",
       "      --bg-color: #E8F0FE;\n",
       "      --fill-color: #1967D2;\n",
       "      --hover-bg-color: #E2EBFA;\n",
       "      --hover-fill-color: #174EA6;\n",
       "      --disabled-fill-color: #AAA;\n",
       "      --disabled-bg-color: #DDD;\n",
       "  }\n",
       "\n",
       "  [theme=dark] .colab-df-quickchart {\n",
       "      --bg-color: #3B4455;\n",
       "      --fill-color: #D2E3FC;\n",
       "      --hover-bg-color: #434B5C;\n",
       "      --hover-fill-color: #FFFFFF;\n",
       "      --disabled-bg-color: #3B4455;\n",
       "      --disabled-fill-color: #666;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart {\n",
       "    background-color: var(--bg-color);\n",
       "    border: none;\n",
       "    border-radius: 50%;\n",
       "    cursor: pointer;\n",
       "    display: none;\n",
       "    fill: var(--fill-color);\n",
       "    height: 32px;\n",
       "    padding: 0;\n",
       "    width: 32px;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart:hover {\n",
       "    background-color: var(--hover-bg-color);\n",
       "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "    fill: var(--button-hover-fill-color);\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart-complete:disabled,\n",
       "  .colab-df-quickchart-complete:disabled:hover {\n",
       "    background-color: var(--disabled-bg-color);\n",
       "    fill: var(--disabled-fill-color);\n",
       "    box-shadow: none;\n",
       "  }\n",
       "\n",
       "  .colab-df-spinner {\n",
       "    border: 2px solid var(--fill-color);\n",
       "    border-color: transparent;\n",
       "    border-bottom-color: var(--fill-color);\n",
       "    animation:\n",
       "      spin 1s steps(1) infinite;\n",
       "  }\n",
       "\n",
       "  @keyframes spin {\n",
       "    0% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "      border-left-color: var(--fill-color);\n",
       "    }\n",
       "    20% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    30% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    40% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    60% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    80% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "    90% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "  }\n",
       "</style>\n",
       "\n",
       "  <script>\n",
       "    async function quickchart(key) {\n",
       "      const quickchartButtonEl =\n",
       "        document.querySelector('#' + key + ' button');\n",
       "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
       "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
       "      try {\n",
       "        const charts = await google.colab.kernel.invokeFunction(\n",
       "            'suggestCharts', [key], {});\n",
       "      } catch (error) {\n",
       "        console.error('Error during call to suggestCharts:', error);\n",
       "      }\n",
       "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
       "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
       "    }\n",
       "    (() => {\n",
       "      let quickchartButtonEl =\n",
       "        document.querySelector('#df-1a3bfaba-0225-4b93-aea9-0a0d48f10bc2 button');\n",
       "      quickchartButtonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "    })();\n",
       "  </script>\n",
       "</div>\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ],
      "text/plain": [
       "   Topic  Count                                  Name  \\\n",
       "0     -1  14980  -1_exposure_metabolic_study_patients   \n",
       "1      0  19258        0_plant_plants_genes_compounds   \n",
       "2      1   9730       1_liver_diabetes_insulin_muscle   \n",
       "3      2   7254             2_cancer_tumor_cells_cell   \n",
       "4      3   6499           3_data_ms_mass_metabolomics   \n",
       "\n",
       "                                      Representation  \\\n",
       "0  [exposure, metabolic, study, patients, metabol...   \n",
       "1  [plant, plants, genes, compounds, stress, grow...   \n",
       "2  [liver, diabetes, insulin, muscle, mice, exerc...   \n",
       "3  [cancer, tumor, cells, cell, breast, cancer ce...   \n",
       "4  [data, ms, mass, metabolomics, sample, method,...   \n",
       "\n",
       "                                         GPT-4o  \\\n",
       "0       [Metabolomics in Disease and Treatment]   \n",
       "1            [Plant Stress Response Mechanisms]   \n",
       "2        [Metabolic Profiles and Dysregulation]   \n",
       "3    [Cancer Metabolism and Therapy Resistance]   \n",
       "4  [Metabolomics Data Analysis and Integration]   \n",
       "\n",
       "                                 Representative_Docs  \n",
       "0  [Background: In numerous studies based predomi...  \n",
       "1  [To investigate differences in fresh leaves of...  \n",
       "2  [Fenugreek is a well-known medicinal plant use...  \n",
       "3  [Cancer cells rewire the metabolic processes b...  \n",
       "4  [Although metabolomics data acquisition and an...  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show topics\n",
    "topic_model.get_topic_info().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "z1ROXU4-l91a"
   },
   "source": [
    "**Generate custom labels**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "Aa9E--ChmD4C"
   },
   "outputs": [],
   "source": [
    "chatgpt_topic_labels = {topic: \" | \".join(list(zip(*values))[0]) for topic, values in topic_model.topic_aspects_[\"GPT-4o\"].items()}\n",
    "chatgpt_topic_labels[-1] = \"Outlier Topic\"\n",
    "topic_model.set_topic_labels(chatgpt_topic_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "executionInfo": {
     "elapsed": 355,
     "status": "ok",
     "timestamp": 1728498661065,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "MIbWXNTtmDvC",
    "outputId": "55906beb-2dc0-4604-df33-09c56712ab54"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "summary": "{\n  \"name\": \"topic_model\",\n  \"rows\": 21,\n  \"fields\": [\n    {\n      \"column\": \"Topic\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 6,\n        \"min\": -1,\n        \"max\": 19,\n        \"num_unique_values\": 21,\n        \"samples\": [\n          -1,\n          16,\n          14\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Count\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 5115,\n        \"min\": 503,\n        \"max\": 19258,\n        \"num_unique_values\": 21,\n        \"samples\": [\n          14980,\n          656,\n          776\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Name\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 21,\n        \"samples\": [\n          \"-1_exposure_metabolic_study_patients\",\n          \"16_parasite_parasites_infection_host\",\n          \"14_covid_covid 19_19_cov\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"CustomName\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 21,\n        \"samples\": [\n          \"Outlier Topic\",\n          \"Host-Parasite Metabolic Interactions\",\n          \"COVID-19 Metabolomic and Immune Responses\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Representation\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"GPT-4o\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Representative_Docs\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}",
       "type": "dataframe"
      },
      "text/html": [
       "\n",
       "  <div id=\"df-7d369cac-8431-4256-88f3-f7bc83078058\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Count</th>\n",
       "      <th>Name</th>\n",
       "      <th>CustomName</th>\n",
       "      <th>Representation</th>\n",
       "      <th>GPT-4o</th>\n",
       "      <th>Representative_Docs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "      <td>14980</td>\n",
       "      <td>-1_exposure_metabolic_study_patients</td>\n",
       "      <td>Outlier Topic</td>\n",
       "      <td>[exposure, metabolic, study, patients, metabol...</td>\n",
       "      <td>[Metabolomics in Disease and Treatment]</td>\n",
       "      <td>[Background: In numerous studies based predomi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>19258</td>\n",
       "      <td>0_plant_plants_genes_compounds</td>\n",
       "      <td>Plant Stress Response Mechanisms</td>\n",
       "      <td>[plant, plants, genes, compounds, stress, grow...</td>\n",
       "      <td>[Plant Stress Response Mechanisms]</td>\n",
       "      <td>[To investigate differences in fresh leaves of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>9730</td>\n",
       "      <td>1_liver_diabetes_insulin_muscle</td>\n",
       "      <td>Metabolic Profiles and Dysregulation</td>\n",
       "      <td>[liver, diabetes, insulin, muscle, mice, exerc...</td>\n",
       "      <td>[Metabolic Profiles and Dysregulation]</td>\n",
       "      <td>[Fenugreek is a well-known medicinal plant use...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>7254</td>\n",
       "      <td>2_cancer_tumor_cells_cell</td>\n",
       "      <td>Cancer Metabolism and Therapy Resistance</td>\n",
       "      <td>[cancer, tumor, cells, cell, breast, cancer ce...</td>\n",
       "      <td>[Cancer Metabolism and Therapy Resistance]</td>\n",
       "      <td>[Cancer cells rewire the metabolic processes b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>6499</td>\n",
       "      <td>3_data_ms_mass_metabolomics</td>\n",
       "      <td>Metabolomics Data Analysis and Integration</td>\n",
       "      <td>[data, ms, mass, metabolomics, sample, method,...</td>\n",
       "      <td>[Metabolomics Data Analysis and Integration]</td>\n",
       "      <td>[Although metabolomics data acquisition and an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>4313</td>\n",
       "      <td>4_gut_microbiota_gut microbiota_microbiome</td>\n",
       "      <td>Gut Microbiota and Metabolomic Interactions</td>\n",
       "      <td>[gut, microbiota, gut microbiota, microbiome, ...</td>\n",
       "      <td>[Gut Microbiota and Metabolomic Interactions]</td>\n",
       "      <td>[Given the high and increasing prevalence of o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5</td>\n",
       "      <td>2723</td>\n",
       "      <td>5_ad_brain_alzheimer_alzheimer disease</td>\n",
       "      <td>Metabolomics in Neurodegenerative Disorders</td>\n",
       "      <td>[ad, brain, alzheimer, alzheimer disease, pd, ...</td>\n",
       "      <td>[Metabolomics in Neurodegenerative Disorders]</td>\n",
       "      <td>[Alzheimer's disease (AD) is the most common c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6</td>\n",
       "      <td>2389</td>\n",
       "      <td>6_fish_zebrafish_exposure_aquatic</td>\n",
       "      <td>Environmental Toxicology and Metabolism</td>\n",
       "      <td>[fish, zebrafish, exposure, aquatic, exposed, ...</td>\n",
       "      <td>[Environmental Toxicology and Metabolism]</td>\n",
       "      <td>[Microplastics (MPs) pollution has been recogn...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7</td>\n",
       "      <td>1971</td>\n",
       "      <td>7_milk_meat_dairy_cattle</td>\n",
       "      <td>Metabolomics in Animal Nutrition</td>\n",
       "      <td>[milk, meat, dairy, cattle, 05, feed, lactatio...</td>\n",
       "      <td>[Metabolomics in Animal Nutrition]</td>\n",
       "      <td>[Ruminants account for a relatively large shar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>8</td>\n",
       "      <td>1661</td>\n",
       "      <td>8_depression_sleep_circadian_schizophrenia</td>\n",
       "      <td>Microbiota-Gut-Brain Axis Interactions</td>\n",
       "      <td>[depression, sleep, circadian, schizophrenia, ...</td>\n",
       "      <td>[Microbiota-Gut-Brain Axis Interactions]</td>\n",
       "      <td>[Prenatal stress (PS) increases offspring susc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9</td>\n",
       "      <td>1479</td>\n",
       "      <td>9_kidney_renal_ckd_kidney disease</td>\n",
       "      <td>Kidney Disease Metabolomics and Biomarkers</td>\n",
       "      <td>[kidney, renal, ckd, kidney disease, chronic k...</td>\n",
       "      <td>[Kidney Disease Metabolomics and Biomarkers]</td>\n",
       "      <td>[Diabetic kidney disease (DKD) is the leading ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>10</td>\n",
       "      <td>1318</td>\n",
       "      <td>10_bone_ra_mscs_oa</td>\n",
       "      <td>Metabolomics in Rheumatoid Arthritis</td>\n",
       "      <td>[bone, ra, mscs, oa, arthritis, secretome, ost...</td>\n",
       "      <td>[Metabolomics in Rheumatoid Arthritis]</td>\n",
       "      <td>[Rheumatoid arthritis (RA) is a rheumatic dise...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>11</td>\n",
       "      <td>1238</td>\n",
       "      <td>11_maternal_pregnancy_fetal_gestational</td>\n",
       "      <td>Maternal Metabolomic Changes in Pregnancy</td>\n",
       "      <td>[maternal, pregnancy, fetal, gestational, offs...</td>\n",
       "      <td>[Maternal Metabolomic Changes in Pregnancy]</td>\n",
       "      <td>[Abnormal maternal lipid profiles, a hallmark ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>12</td>\n",
       "      <td>1186</td>\n",
       "      <td>12_asthma_lung_pulmonary_airway</td>\n",
       "      <td>Lung Disease and Metabolic Dynamics</td>\n",
       "      <td>[asthma, lung, pulmonary, airway, allergic, re...</td>\n",
       "      <td>[Lung Disease and Metabolic Dynamics]</td>\n",
       "      <td>[Breathomics, the multidimensional molecular a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>13</td>\n",
       "      <td>779</td>\n",
       "      <td>13_nmr_spectra_spectroscopy_nmr spectra</td>\n",
       "      <td>NMR Spectroscopy Innovations in Metabolomics</td>\n",
       "      <td>[nmr, spectra, spectroscopy, nmr spectra, nmr ...</td>\n",
       "      <td>[NMR Spectroscopy Innovations in Metabolomics]</td>\n",
       "      <td>[Nuclear magnetic resonance (NMR) and Mass Spe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>14</td>\n",
       "      <td>776</td>\n",
       "      <td>14_covid_covid 19_19_cov</td>\n",
       "      <td>COVID-19 Metabolomic and Immune Responses</td>\n",
       "      <td>[covid, covid 19, 19, cov, sars, sars cov, inf...</td>\n",
       "      <td>[COVID-19 Metabolomic and Immune Responses]</td>\n",
       "      <td>[Coronavirus disease 2019 (COVID-19) pandemic ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>15</td>\n",
       "      <td>687</td>\n",
       "      <td>15_lipidomics_lipid_lipids_lipid species</td>\n",
       "      <td>Lipid Profiling Techniques</td>\n",
       "      <td>[lipidomics, lipid, lipids, lipid species, lip...</td>\n",
       "      <td>[Lipid Profiling Techniques]</td>\n",
       "      <td>[Lipids are an important class of biomolecules...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>16</td>\n",
       "      <td>656</td>\n",
       "      <td>16_parasite_parasites_infection_host</td>\n",
       "      <td>Host-Parasite Metabolic Interactions</td>\n",
       "      <td>[parasite, parasites, infection, host, parasit...</td>\n",
       "      <td>[Host-Parasite Metabolic Interactions]</td>\n",
       "      <td>[The hygiene hypothesis proposes that decrease...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>17</td>\n",
       "      <td>652</td>\n",
       "      <td>17_sperm_embryo_oocyte_follicular</td>\n",
       "      <td>Male Fertility and Reproductive Metabolomics</td>\n",
       "      <td>[sperm, embryo, oocyte, follicular, embryos, i...</td>\n",
       "      <td>[Male Fertility and Reproductive Metabolomics]</td>\n",
       "      <td>[The identification of markers capable of eval...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>18</td>\n",
       "      <td>604</td>\n",
       "      <td>18_retinal_retina_dr_ocular</td>\n",
       "      <td>Ocular Metabolomics and Disease Mechanisms</td>\n",
       "      <td>[retinal, retina, dr, ocular, eye, retinopathy...</td>\n",
       "      <td>[Ocular Metabolomics and Disease Mechanisms]</td>\n",
       "      <td>[Visual deficits are a common concern among su...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>19</td>\n",
       "      <td>503</td>\n",
       "      <td>19_oral_saliva_salivary_dental</td>\n",
       "      <td>Salivary Metabolomics in Oral Health</td>\n",
       "      <td>[oral, saliva, salivary, dental, saliva sample...</td>\n",
       "      <td>[Salivary Metabolomics in Oral Health]</td>\n",
       "      <td>[This communication represents Part III of our...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-7d369cac-8431-4256-88f3-f7bc83078058')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-7d369cac-8431-4256-88f3-f7bc83078058 button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-7d369cac-8431-4256-88f3-f7bc83078058');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "<div id=\"df-7469c3fd-5272-4bc0-9ced-0c00e134fb37\">\n",
       "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-7469c3fd-5272-4bc0-9ced-0c00e134fb37')\"\n",
       "            title=\"Suggest charts\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "     width=\"24px\">\n",
       "    <g>\n",
       "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
       "    </g>\n",
       "</svg>\n",
       "  </button>\n",
       "\n",
       "<style>\n",
       "  .colab-df-quickchart {\n",
       "      --bg-color: #E8F0FE;\n",
       "      --fill-color: #1967D2;\n",
       "      --hover-bg-color: #E2EBFA;\n",
       "      --hover-fill-color: #174EA6;\n",
       "      --disabled-fill-color: #AAA;\n",
       "      --disabled-bg-color: #DDD;\n",
       "  }\n",
       "\n",
       "  [theme=dark] .colab-df-quickchart {\n",
       "      --bg-color: #3B4455;\n",
       "      --fill-color: #D2E3FC;\n",
       "      --hover-bg-color: #434B5C;\n",
       "      --hover-fill-color: #FFFFFF;\n",
       "      --disabled-bg-color: #3B4455;\n",
       "      --disabled-fill-color: #666;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart {\n",
       "    background-color: var(--bg-color);\n",
       "    border: none;\n",
       "    border-radius: 50%;\n",
       "    cursor: pointer;\n",
       "    display: none;\n",
       "    fill: var(--fill-color);\n",
       "    height: 32px;\n",
       "    padding: 0;\n",
       "    width: 32px;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart:hover {\n",
       "    background-color: var(--hover-bg-color);\n",
       "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "    fill: var(--button-hover-fill-color);\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart-complete:disabled,\n",
       "  .colab-df-quickchart-complete:disabled:hover {\n",
       "    background-color: var(--disabled-bg-color);\n",
       "    fill: var(--disabled-fill-color);\n",
       "    box-shadow: none;\n",
       "  }\n",
       "\n",
       "  .colab-df-spinner {\n",
       "    border: 2px solid var(--fill-color);\n",
       "    border-color: transparent;\n",
       "    border-bottom-color: var(--fill-color);\n",
       "    animation:\n",
       "      spin 1s steps(1) infinite;\n",
       "  }\n",
       "\n",
       "  @keyframes spin {\n",
       "    0% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "      border-left-color: var(--fill-color);\n",
       "    }\n",
       "    20% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    30% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    40% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    60% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    80% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "    90% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "  }\n",
       "</style>\n",
       "\n",
       "  <script>\n",
       "    async function quickchart(key) {\n",
       "      const quickchartButtonEl =\n",
       "        document.querySelector('#' + key + ' button');\n",
       "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
       "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
       "      try {\n",
       "        const charts = await google.colab.kernel.invokeFunction(\n",
       "            'suggestCharts', [key], {});\n",
       "      } catch (error) {\n",
       "        console.error('Error during call to suggestCharts:', error);\n",
       "      }\n",
       "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
       "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
       "    }\n",
       "    (() => {\n",
       "      let quickchartButtonEl =\n",
       "        document.querySelector('#df-7469c3fd-5272-4bc0-9ced-0c00e134fb37 button');\n",
       "      quickchartButtonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "    })();\n",
       "  </script>\n",
       "</div>\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ],
      "text/plain": [
       "    Topic  Count                                        Name  \\\n",
       "0      -1  14980        -1_exposure_metabolic_study_patients   \n",
       "1       0  19258              0_plant_plants_genes_compounds   \n",
       "2       1   9730             1_liver_diabetes_insulin_muscle   \n",
       "3       2   7254                   2_cancer_tumor_cells_cell   \n",
       "4       3   6499                 3_data_ms_mass_metabolomics   \n",
       "5       4   4313  4_gut_microbiota_gut microbiota_microbiome   \n",
       "6       5   2723      5_ad_brain_alzheimer_alzheimer disease   \n",
       "7       6   2389           6_fish_zebrafish_exposure_aquatic   \n",
       "8       7   1971                    7_milk_meat_dairy_cattle   \n",
       "9       8   1661  8_depression_sleep_circadian_schizophrenia   \n",
       "10      9   1479           9_kidney_renal_ckd_kidney disease   \n",
       "11     10   1318                          10_bone_ra_mscs_oa   \n",
       "12     11   1238     11_maternal_pregnancy_fetal_gestational   \n",
       "13     12   1186             12_asthma_lung_pulmonary_airway   \n",
       "14     13    779     13_nmr_spectra_spectroscopy_nmr spectra   \n",
       "15     14    776                    14_covid_covid 19_19_cov   \n",
       "16     15    687    15_lipidomics_lipid_lipids_lipid species   \n",
       "17     16    656        16_parasite_parasites_infection_host   \n",
       "18     17    652           17_sperm_embryo_oocyte_follicular   \n",
       "19     18    604                 18_retinal_retina_dr_ocular   \n",
       "20     19    503              19_oral_saliva_salivary_dental   \n",
       "\n",
       "                                      CustomName  \\\n",
       "0                                  Outlier Topic   \n",
       "1               Plant Stress Response Mechanisms   \n",
       "2           Metabolic Profiles and Dysregulation   \n",
       "3       Cancer Metabolism and Therapy Resistance   \n",
       "4     Metabolomics Data Analysis and Integration   \n",
       "5    Gut Microbiota and Metabolomic Interactions   \n",
       "6    Metabolomics in Neurodegenerative Disorders   \n",
       "7        Environmental Toxicology and Metabolism   \n",
       "8               Metabolomics in Animal Nutrition   \n",
       "9         Microbiota-Gut-Brain Axis Interactions   \n",
       "10    Kidney Disease Metabolomics and Biomarkers   \n",
       "11          Metabolomics in Rheumatoid Arthritis   \n",
       "12     Maternal Metabolomic Changes in Pregnancy   \n",
       "13           Lung Disease and Metabolic Dynamics   \n",
       "14  NMR Spectroscopy Innovations in Metabolomics   \n",
       "15     COVID-19 Metabolomic and Immune Responses   \n",
       "16                    Lipid Profiling Techniques   \n",
       "17          Host-Parasite Metabolic Interactions   \n",
       "18  Male Fertility and Reproductive Metabolomics   \n",
       "19    Ocular Metabolomics and Disease Mechanisms   \n",
       "20          Salivary Metabolomics in Oral Health   \n",
       "\n",
       "                                       Representation  \\\n",
       "0   [exposure, metabolic, study, patients, metabol...   \n",
       "1   [plant, plants, genes, compounds, stress, grow...   \n",
       "2   [liver, diabetes, insulin, muscle, mice, exerc...   \n",
       "3   [cancer, tumor, cells, cell, breast, cancer ce...   \n",
       "4   [data, ms, mass, metabolomics, sample, method,...   \n",
       "5   [gut, microbiota, gut microbiota, microbiome, ...   \n",
       "6   [ad, brain, alzheimer, alzheimer disease, pd, ...   \n",
       "7   [fish, zebrafish, exposure, aquatic, exposed, ...   \n",
       "8   [milk, meat, dairy, cattle, 05, feed, lactatio...   \n",
       "9   [depression, sleep, circadian, schizophrenia, ...   \n",
       "10  [kidney, renal, ckd, kidney disease, chronic k...   \n",
       "11  [bone, ra, mscs, oa, arthritis, secretome, ost...   \n",
       "12  [maternal, pregnancy, fetal, gestational, offs...   \n",
       "13  [asthma, lung, pulmonary, airway, allergic, re...   \n",
       "14  [nmr, spectra, spectroscopy, nmr spectra, nmr ...   \n",
       "15  [covid, covid 19, 19, cov, sars, sars cov, inf...   \n",
       "16  [lipidomics, lipid, lipids, lipid species, lip...   \n",
       "17  [parasite, parasites, infection, host, parasit...   \n",
       "18  [sperm, embryo, oocyte, follicular, embryos, i...   \n",
       "19  [retinal, retina, dr, ocular, eye, retinopathy...   \n",
       "20  [oral, saliva, salivary, dental, saliva sample...   \n",
       "\n",
       "                                            GPT-4o  \\\n",
       "0          [Metabolomics in Disease and Treatment]   \n",
       "1               [Plant Stress Response Mechanisms]   \n",
       "2           [Metabolic Profiles and Dysregulation]   \n",
       "3       [Cancer Metabolism and Therapy Resistance]   \n",
       "4     [Metabolomics Data Analysis and Integration]   \n",
       "5    [Gut Microbiota and Metabolomic Interactions]   \n",
       "6    [Metabolomics in Neurodegenerative Disorders]   \n",
       "7        [Environmental Toxicology and Metabolism]   \n",
       "8               [Metabolomics in Animal Nutrition]   \n",
       "9         [Microbiota-Gut-Brain Axis Interactions]   \n",
       "10    [Kidney Disease Metabolomics and Biomarkers]   \n",
       "11          [Metabolomics in Rheumatoid Arthritis]   \n",
       "12     [Maternal Metabolomic Changes in Pregnancy]   \n",
       "13           [Lung Disease and Metabolic Dynamics]   \n",
       "14  [NMR Spectroscopy Innovations in Metabolomics]   \n",
       "15     [COVID-19 Metabolomic and Immune Responses]   \n",
       "16                    [Lipid Profiling Techniques]   \n",
       "17          [Host-Parasite Metabolic Interactions]   \n",
       "18  [Male Fertility and Reproductive Metabolomics]   \n",
       "19    [Ocular Metabolomics and Disease Mechanisms]   \n",
       "20          [Salivary Metabolomics in Oral Health]   \n",
       "\n",
       "                                  Representative_Docs  \n",
       "0   [Background: In numerous studies based predomi...  \n",
       "1   [To investigate differences in fresh leaves of...  \n",
       "2   [Fenugreek is a well-known medicinal plant use...  \n",
       "3   [Cancer cells rewire the metabolic processes b...  \n",
       "4   [Although metabolomics data acquisition and an...  \n",
       "5   [Given the high and increasing prevalence of o...  \n",
       "6   [Alzheimer's disease (AD) is the most common c...  \n",
       "7   [Microplastics (MPs) pollution has been recogn...  \n",
       "8   [Ruminants account for a relatively large shar...  \n",
       "9   [Prenatal stress (PS) increases offspring susc...  \n",
       "10  [Diabetic kidney disease (DKD) is the leading ...  \n",
       "11  [Rheumatoid arthritis (RA) is a rheumatic dise...  \n",
       "12  [Abnormal maternal lipid profiles, a hallmark ...  \n",
       "13  [Breathomics, the multidimensional molecular a...  \n",
       "14  [Nuclear magnetic resonance (NMR) and Mass Spe...  \n",
       "15  [Coronavirus disease 2019 (COVID-19) pandemic ...  \n",
       "16  [Lipids are an important class of biomolecules...  \n",
       "17  [The hygiene hypothesis proposes that decrease...  \n",
       "18  [The identification of markers capable of eval...  \n",
       "19  [Visual deficits are a common concern among su...  \n",
       "20  [This communication represents Part III of our...  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show topics\n",
    "topic_model.get_topic_info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "JWEYXHew1AwS"
   },
   "source": [
    "Here is the topic model for a specific cluster. <br>\n",
    "`Main` represents the c-TF-IDF topic representation outputs, in this case indicating the top 10 topic word. `OpenAI` and `Llama` is used for topic distillation.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 319,
     "status": "ok",
     "timestamp": 1728498724507,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "6OhXo9NBfWSD",
    "outputId": "2705687a-3737-47be-b484-798ec25f922f"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Main': [('cancer', 0.37852654333747204),\n",
       "  ('tumor', 0.312547535360719),\n",
       "  ('cells', 0.305661962531353),\n",
       "  ('cell', 0.2660296334483482),\n",
       "  ('breast', 0.247039976881673),\n",
       "  ('cancer cells', 0.24132808303516712),\n",
       "  ('breast cancer', 0.23847543598042675),\n",
       "  ('tumors', 0.23433026384303857),\n",
       "  ('mitochondrial', 0.22064409809677088),\n",
       "  ('patients', 0.2173006254780799)],\n",
       " 'GPT-4o': [('Cancer Metabolism and Therapy Resistance', 1)]}"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_model.get_topic(2, full=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "1rCuEpGXf8WR",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Outlier Reduction - Skipped."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "FrwaNCDwhUCf"
   },
   "source": [
    "I have attempted using outlier reduction, based on my observation, there tends to be topic contamination, so this process were skipped."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "79hTC_WEgmJV"
   },
   "source": [
    "By default, HDBSCAN generates outliers which is a helpful mechanic in creating accurate topic representations. However, you might want to assign every single document to a topic. We can use `.reduce_outliers` to map some or all outliers to a topic. Here we use the `tf-idf` strategy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "rJ7TnRvZrbaC"
   },
   "outputs": [],
   "source": [
    "# Use the \"c-TF-IDF\" strategy with a threshold\n",
    "new_topics = topic_model.reduce_outliers(abstracts,\n",
    "                                         topics ,\n",
    "                                         strategy=\"c-tf-idf\",\n",
    "                                         threshold=0.12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "N0dBtLgdtRUP"
   },
   "source": [
    "The outlier reduction strategy above has reduced the outlier documents to 8,261 papers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 3,
     "status": "ok",
     "timestamp": 1728497850294,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "FmyNkuS5sxVb",
    "outputId": "8c262755-8ea7-4250-c6f0-4e13c0704c06"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18367\n",
      "8261\n"
     ]
    }
   ],
   "source": [
    "print(topics.count(-1))\n",
    "print(new_topics.count(-1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "E6Unq3vohvt0"
   },
   "source": [
    "**Note:** updating the topics this way may lead to errors if topic reduction or topic merging techniques are used afterwards. The reason for this is that when you assign a -1 document to topic 1 and another -1 document to topic 2, it is unclear how you map the -1 documents. Is it matched to topic 1 or 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 148439,
     "status": "ok",
     "timestamp": 1728498007553,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "GXUxTgRUxmY3",
    "outputId": "915e92ca-3d97-4a24-cc31-d5ba73747fef"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-09 18:17:39,027 - BERTopic - WARNING: Using a custom list of topic assignments may lead to errors if topic reduction techniques are used afterwards. Make sure that manually assigning topics is the last step in the pipeline.Note that topic embeddings will also be created through weightedc-TF-IDF embeddings instead of centroid embeddings.\n",
      "100%|██████████| 21/21 [02:14<00:00,  6.38s/it]\n"
     ]
    }
   ],
   "source": [
    "# Apply the representation model to rename topics\n",
    "topic_model.update_topics(abstracts,\n",
    "                          topics=new_topics,\n",
    "                          representation_model=representation_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "executionInfo": {
     "elapsed": 407,
     "status": "ok",
     "timestamp": 1728498011257,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "ii1ARyL9iAoc",
    "outputId": "7b6d21d9-9d8f-4cc7-ef56-80fe35da360a"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "summary": "{\n  \"name\": \"topic_model\",\n  \"rows\": 21,\n  \"fields\": [\n    {\n      \"column\": \"Topic\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 6,\n        \"min\": -1,\n        \"max\": 19,\n        \"num_unique_values\": 21,\n        \"samples\": [\n          -1,\n          16,\n          14\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Count\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 4516,\n        \"min\": 594,\n        \"max\": 19065,\n        \"num_unique_values\": 21,\n        \"samples\": [\n          8261,\n          792,\n          1042\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Name\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 21,\n        \"samples\": [\n          \"-1_of_the_and_in\",\n          \"16_meat_the_and_of\",\n          \"14_lipid_of_the_and\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Representation\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"GPT-4o\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Representative_Docs\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}",
       "type": "dataframe"
      },
      "text/html": [
       "\n",
       "  <div id=\"df-27b95e4b-b607-44df-8154-bf6057594afa\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Count</th>\n",
       "      <th>Name</th>\n",
       "      <th>Representation</th>\n",
       "      <th>GPT-4o</th>\n",
       "      <th>Representative_Docs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "      <td>8261</td>\n",
       "      <td>-1_of_the_and_in</td>\n",
       "      <td>[of, the, and, in, to, is, with, for, that, we]</td>\n",
       "      <td>[Metabolomics in Disease Mechanisms]</td>\n",
       "      <td>[Vitamin D is a steroid hormone precursor that...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>19065</td>\n",
       "      <td>0_the_of_and_in</td>\n",
       "      <td>[the, of, and, in, to, were, that, for, as, by]</td>\n",
       "      <td>[Plant Metabolic Responses to Stress]</td>\n",
       "      <td>[Celery is an important leafy vegetable that c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>9960</td>\n",
       "      <td>1_and_the_of_in</td>\n",
       "      <td>[and, the, of, in, to, with, were, liver, that...</td>\n",
       "      <td>[Metabolomic Insights into Disease States]</td>\n",
       "      <td>[Non-alcoholic steatohepatitis (NASH) is a sev...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>7878</td>\n",
       "      <td>2_the_of_and_to</td>\n",
       "      <td>[the, of, and, to, for, data, in, ms, is, meta...</td>\n",
       "      <td>[Metabolomics in Disease Diagnosis]</td>\n",
       "      <td>[High-resolution mass spectrometry (HRMS)-base...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>7378</td>\n",
       "      <td>3_cancer_of_and_in</td>\n",
       "      <td>[cancer, of, and, in, the, cells, to, cell, tu...</td>\n",
       "      <td>[Cancer Metabolism and Biomarker Discovery]</td>\n",
       "      <td>[Triple-negative breast cancer (TNBC) is the m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>4892</td>\n",
       "      <td>4_gut_the_microbiota_and</td>\n",
       "      <td>[gut, the, microbiota, and, of, in, to, microb...</td>\n",
       "      <td>[Gut Microbiota and Metabolism]</td>\n",
       "      <td>[Background: Recent evidence suggests that the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5</td>\n",
       "      <td>3092</td>\n",
       "      <td>5_and_the_of_in</td>\n",
       "      <td>[and, the, of, in, to, with, covid, 19, for, t...</td>\n",
       "      <td>[Metabolomics in COVID-19 Research]</td>\n",
       "      <td>[Obese patientss with nonalcoholic steatohepat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6</td>\n",
       "      <td>3054</td>\n",
       "      <td>6_the_and_of_in</td>\n",
       "      <td>[the, and, of, in, brain, ad, to, disease, wit...</td>\n",
       "      <td>[Metabolic Biomarkers in Neurodegenerative Dis...</td>\n",
       "      <td>[(1) Background: Alzheimer's disease (AD) is a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7</td>\n",
       "      <td>2979</td>\n",
       "      <td>7_the_and_of_in</td>\n",
       "      <td>[the, and, of, in, to, exposure, were, that, m...</td>\n",
       "      <td>[Environmental Contaminant Impact on Metabolism]</td>\n",
       "      <td>[Copper (Cu) is a micronutrient essential for ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>8</td>\n",
       "      <td>1858</td>\n",
       "      <td>8_kidney_and_of_renal</td>\n",
       "      <td>[kidney, and, of, renal, the, in, to, ckd, wit...</td>\n",
       "      <td>[Metabolomics in Kidney Disease Research]</td>\n",
       "      <td>[Hepatorenal syndrome (HRS) continues to be on...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9</td>\n",
       "      <td>1595</td>\n",
       "      <td>9_and_the_of_in</td>\n",
       "      <td>[and, the, of, in, depression, gut, to, with, ...</td>\n",
       "      <td>[Gut-Brain Axis in Depression]</td>\n",
       "      <td>[Accumulating evidence suggests that gut micro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>10</td>\n",
       "      <td>1605</td>\n",
       "      <td>10_the_and_of_in</td>\n",
       "      <td>[the, and, of, in, to, ra, bone, with, were, m...</td>\n",
       "      <td>[Metabolomics in Rheumatoid Arthritis]</td>\n",
       "      <td>[Bone-marrow mesenchymal stem cells (BM-MSCs) ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>11</td>\n",
       "      <td>1694</td>\n",
       "      <td>11_and_of_the_maternal</td>\n",
       "      <td>[and, of, the, maternal, in, pregnancy, to, wi...</td>\n",
       "      <td>[Maternal Metabolomics and Pregnancy Outcomes]</td>\n",
       "      <td>[Pregnancy has been equated to a \"stress test\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>12</td>\n",
       "      <td>1303</td>\n",
       "      <td>12_milk_and_the_cows</td>\n",
       "      <td>[milk, and, the, cows, were, of, in, to, rumen...</td>\n",
       "      <td>[Rumen Metabolomics and Microbiota Dynamics]</td>\n",
       "      <td>[Ruminants account for a relatively large shar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>13</td>\n",
       "      <td>1219</td>\n",
       "      <td>13_nmr_of_the_and</td>\n",
       "      <td>[nmr, of, the, and, to, for, in, spectroscopy,...</td>\n",
       "      <td>[NMR-Based Metabolomics Advances]</td>\n",
       "      <td>[NMR spectroscopy is a principal tool in metab...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>14</td>\n",
       "      <td>1042</td>\n",
       "      <td>14_lipid_of_the_and</td>\n",
       "      <td>[lipid, of, the, and, lipids, lipidomics, in, ...</td>\n",
       "      <td>[Advancements in Lipidomics Techniques]</td>\n",
       "      <td>[Lipids are an important class of biomolecules...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>15</td>\n",
       "      <td>744</td>\n",
       "      <td>15_sperm_the_of_and</td>\n",
       "      <td>[sperm, the, of, and, in, embryo, to, embryos,...</td>\n",
       "      <td>[Metabolomics in Reproductive Health]</td>\n",
       "      <td>[Cattle induced to ovulate a small, physiologi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>16</td>\n",
       "      <td>792</td>\n",
       "      <td>16_meat_the_and_of</td>\n",
       "      <td>[meat, the, and, of, in, were, to, metabolites...</td>\n",
       "      <td>[Meat Quality and Metabolism Insights]</td>\n",
       "      <td>[Selenium (Se) is an indispensable trace eleme...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>17</td>\n",
       "      <td>789</td>\n",
       "      <td>17_the_of_parasite_and</td>\n",
       "      <td>[the, of, parasite, and, in, to, infection, ho...</td>\n",
       "      <td>[Host-Pathogen Metabolic Interactions]</td>\n",
       "      <td>[Metabolomics refers to top-down systems biolo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>18</td>\n",
       "      <td>594</td>\n",
       "      <td>18_retinal_the_and_in</td>\n",
       "      <td>[retinal, the, and, in, of, to, dr, amd, with,...</td>\n",
       "      <td>[Metabolic Insights into Retinal Disorders]</td>\n",
       "      <td>[Background: Age-related macular degeneration ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>19</td>\n",
       "      <td>862</td>\n",
       "      <td>19_fermentation_the_and_of</td>\n",
       "      <td>[fermentation, the, and, of, were, in, ferment...</td>\n",
       "      <td>[Microbial Fermentation Metabolomics]</td>\n",
       "      <td>[Microbial ecosystems of fermented foods are l...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-27b95e4b-b607-44df-8154-bf6057594afa')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-27b95e4b-b607-44df-8154-bf6057594afa button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-27b95e4b-b607-44df-8154-bf6057594afa');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "<div id=\"df-8690f60a-5609-4e53-925c-58eb1094d863\">\n",
       "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-8690f60a-5609-4e53-925c-58eb1094d863')\"\n",
       "            title=\"Suggest charts\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "     width=\"24px\">\n",
       "    <g>\n",
       "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
       "    </g>\n",
       "</svg>\n",
       "  </button>\n",
       "\n",
       "<style>\n",
       "  .colab-df-quickchart {\n",
       "      --bg-color: #E8F0FE;\n",
       "      --fill-color: #1967D2;\n",
       "      --hover-bg-color: #E2EBFA;\n",
       "      --hover-fill-color: #174EA6;\n",
       "      --disabled-fill-color: #AAA;\n",
       "      --disabled-bg-color: #DDD;\n",
       "  }\n",
       "\n",
       "  [theme=dark] .colab-df-quickchart {\n",
       "      --bg-color: #3B4455;\n",
       "      --fill-color: #D2E3FC;\n",
       "      --hover-bg-color: #434B5C;\n",
       "      --hover-fill-color: #FFFFFF;\n",
       "      --disabled-bg-color: #3B4455;\n",
       "      --disabled-fill-color: #666;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart {\n",
       "    background-color: var(--bg-color);\n",
       "    border: none;\n",
       "    border-radius: 50%;\n",
       "    cursor: pointer;\n",
       "    display: none;\n",
       "    fill: var(--fill-color);\n",
       "    height: 32px;\n",
       "    padding: 0;\n",
       "    width: 32px;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart:hover {\n",
       "    background-color: var(--hover-bg-color);\n",
       "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "    fill: var(--button-hover-fill-color);\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart-complete:disabled,\n",
       "  .colab-df-quickchart-complete:disabled:hover {\n",
       "    background-color: var(--disabled-bg-color);\n",
       "    fill: var(--disabled-fill-color);\n",
       "    box-shadow: none;\n",
       "  }\n",
       "\n",
       "  .colab-df-spinner {\n",
       "    border: 2px solid var(--fill-color);\n",
       "    border-color: transparent;\n",
       "    border-bottom-color: var(--fill-color);\n",
       "    animation:\n",
       "      spin 1s steps(1) infinite;\n",
       "  }\n",
       "\n",
       "  @keyframes spin {\n",
       "    0% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "      border-left-color: var(--fill-color);\n",
       "    }\n",
       "    20% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    30% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    40% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    60% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    80% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "    90% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "  }\n",
       "</style>\n",
       "\n",
       "  <script>\n",
       "    async function quickchart(key) {\n",
       "      const quickchartButtonEl =\n",
       "        document.querySelector('#' + key + ' button');\n",
       "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
       "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
       "      try {\n",
       "        const charts = await google.colab.kernel.invokeFunction(\n",
       "            'suggestCharts', [key], {});\n",
       "      } catch (error) {\n",
       "        console.error('Error during call to suggestCharts:', error);\n",
       "      }\n",
       "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
       "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
       "    }\n",
       "    (() => {\n",
       "      let quickchartButtonEl =\n",
       "        document.querySelector('#df-8690f60a-5609-4e53-925c-58eb1094d863 button');\n",
       "      quickchartButtonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "    })();\n",
       "  </script>\n",
       "</div>\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ],
      "text/plain": [
       "    Topic  Count                        Name  \\\n",
       "0      -1   8261            -1_of_the_and_in   \n",
       "1       0  19065             0_the_of_and_in   \n",
       "2       1   9960             1_and_the_of_in   \n",
       "3       2   7878             2_the_of_and_to   \n",
       "4       3   7378          3_cancer_of_and_in   \n",
       "5       4   4892    4_gut_the_microbiota_and   \n",
       "6       5   3092             5_and_the_of_in   \n",
       "7       6   3054             6_the_and_of_in   \n",
       "8       7   2979             7_the_and_of_in   \n",
       "9       8   1858       8_kidney_and_of_renal   \n",
       "10      9   1595             9_and_the_of_in   \n",
       "11     10   1605            10_the_and_of_in   \n",
       "12     11   1694      11_and_of_the_maternal   \n",
       "13     12   1303        12_milk_and_the_cows   \n",
       "14     13   1219           13_nmr_of_the_and   \n",
       "15     14   1042         14_lipid_of_the_and   \n",
       "16     15    744         15_sperm_the_of_and   \n",
       "17     16    792          16_meat_the_and_of   \n",
       "18     17    789      17_the_of_parasite_and   \n",
       "19     18    594       18_retinal_the_and_in   \n",
       "20     19    862  19_fermentation_the_and_of   \n",
       "\n",
       "                                       Representation  \\\n",
       "0     [of, the, and, in, to, is, with, for, that, we]   \n",
       "1     [the, of, and, in, to, were, that, for, as, by]   \n",
       "2   [and, the, of, in, to, with, were, liver, that...   \n",
       "3   [the, of, and, to, for, data, in, ms, is, meta...   \n",
       "4   [cancer, of, and, in, the, cells, to, cell, tu...   \n",
       "5   [gut, the, microbiota, and, of, in, to, microb...   \n",
       "6   [and, the, of, in, to, with, covid, 19, for, t...   \n",
       "7   [the, and, of, in, brain, ad, to, disease, wit...   \n",
       "8   [the, and, of, in, to, exposure, were, that, m...   \n",
       "9   [kidney, and, of, renal, the, in, to, ckd, wit...   \n",
       "10  [and, the, of, in, depression, gut, to, with, ...   \n",
       "11  [the, and, of, in, to, ra, bone, with, were, m...   \n",
       "12  [and, of, the, maternal, in, pregnancy, to, wi...   \n",
       "13  [milk, and, the, cows, were, of, in, to, rumen...   \n",
       "14  [nmr, of, the, and, to, for, in, spectroscopy,...   \n",
       "15  [lipid, of, the, and, lipids, lipidomics, in, ...   \n",
       "16  [sperm, the, of, and, in, embryo, to, embryos,...   \n",
       "17  [meat, the, and, of, in, were, to, metabolites...   \n",
       "18  [the, of, parasite, and, in, to, infection, ho...   \n",
       "19  [retinal, the, and, in, of, to, dr, amd, with,...   \n",
       "20  [fermentation, the, and, of, were, in, ferment...   \n",
       "\n",
       "                                               GPT-4o  \\\n",
       "0                [Metabolomics in Disease Mechanisms]   \n",
       "1               [Plant Metabolic Responses to Stress]   \n",
       "2          [Metabolomic Insights into Disease States]   \n",
       "3                 [Metabolomics in Disease Diagnosis]   \n",
       "4         [Cancer Metabolism and Biomarker Discovery]   \n",
       "5                     [Gut Microbiota and Metabolism]   \n",
       "6                 [Metabolomics in COVID-19 Research]   \n",
       "7   [Metabolic Biomarkers in Neurodegenerative Dis...   \n",
       "8    [Environmental Contaminant Impact on Metabolism]   \n",
       "9           [Metabolomics in Kidney Disease Research]   \n",
       "10                     [Gut-Brain Axis in Depression]   \n",
       "11             [Metabolomics in Rheumatoid Arthritis]   \n",
       "12     [Maternal Metabolomics and Pregnancy Outcomes]   \n",
       "13       [Rumen Metabolomics and Microbiota Dynamics]   \n",
       "14                  [NMR-Based Metabolomics Advances]   \n",
       "15            [Advancements in Lipidomics Techniques]   \n",
       "16              [Metabolomics in Reproductive Health]   \n",
       "17             [Meat Quality and Metabolism Insights]   \n",
       "18             [Host-Pathogen Metabolic Interactions]   \n",
       "19        [Metabolic Insights into Retinal Disorders]   \n",
       "20              [Microbial Fermentation Metabolomics]   \n",
       "\n",
       "                                  Representative_Docs  \n",
       "0   [Vitamin D is a steroid hormone precursor that...  \n",
       "1   [Celery is an important leafy vegetable that c...  \n",
       "2   [Non-alcoholic steatohepatitis (NASH) is a sev...  \n",
       "3   [High-resolution mass spectrometry (HRMS)-base...  \n",
       "4   [Triple-negative breast cancer (TNBC) is the m...  \n",
       "5   [Background: Recent evidence suggests that the...  \n",
       "6   [Obese patientss with nonalcoholic steatohepat...  \n",
       "7   [(1) Background: Alzheimer's disease (AD) is a...  \n",
       "8   [Copper (Cu) is a micronutrient essential for ...  \n",
       "9   [Hepatorenal syndrome (HRS) continues to be on...  \n",
       "10  [Accumulating evidence suggests that gut micro...  \n",
       "11  [Bone-marrow mesenchymal stem cells (BM-MSCs) ...  \n",
       "12  [Pregnancy has been equated to a \"stress test\"...  \n",
       "13  [Ruminants account for a relatively large shar...  \n",
       "14  [NMR spectroscopy is a principal tool in metab...  \n",
       "15  [Lipids are an important class of biomolecules...  \n",
       "16  [Cattle induced to ovulate a small, physiologi...  \n",
       "17  [Selenium (Se) is an indispensable trace eleme...  \n",
       "18  [Metabolomics refers to top-down systems biolo...  \n",
       "19  [Background: Age-related macular degeneration ...  \n",
       "20  [Microbial ecosystems of fermented foods are l...  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show udated topics\n",
    "topic_model.get_topic_info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 266,
     "status": "ok",
     "timestamp": 1728486260514,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "aKHwRzxVvU5G",
    "outputId": "38f14ea9-cb1e-4a75-d126-a1a214cb3301"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Main': [('the', 0.028779946662807827),\n",
       "  ('of', 0.02864578100666034),\n",
       "  ('and', 0.02448622292237831),\n",
       "  ('to', 0.02432008961573552),\n",
       "  ('for', 0.0241340138406972),\n",
       "  ('data', 0.02266133022556849),\n",
       "  ('in', 0.0201851148834008),\n",
       "  ('ms', 0.019470664827897718),\n",
       "  ('is', 0.016727441583856882),\n",
       "  ('mass', 0.015428174802613718)],\n",
       " 'GPT-4o': [('Mass Spectrometry-Based Metabolomics Analysis', 1)]}"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_model.get_topic(1, full=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "xnT8aq0SzXzv"
   },
   "source": [
    "Update custom names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "eJyQxA4Yxe64"
   },
   "outputs": [],
   "source": [
    "chatgpt_topic_labels = {topic: \" | \".join(list(zip(*values))[0]) for topic, values in topic_model.topic_aspects_[\"GPT-4o\"].items()}\n",
    "#chatgpt_topic_labels[-1] = \"Outlier Topic\"\n",
    "topic_model.set_topic_labels(chatgpt_topic_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 521
    },
    "executionInfo": {
     "elapsed": 208,
     "status": "ok",
     "timestamp": 1728486236330,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "32AJOojgxel-",
    "outputId": "2e4fc0b6-9c4e-4db8-f1a8-f68ddc32a058"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "summary": "{\n  \"name\": \"topic_model\",\n  \"rows\": 5,\n  \"fields\": [\n    {\n      \"column\": \"Topic\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 1,\n        \"min\": -1,\n        \"max\": 3,\n        \"num_unique_values\": 5,\n        \"samples\": [\n          0,\n          3,\n          1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Count\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 6666,\n        \"min\": 2632,\n        \"max\": 19955,\n        \"num_unique_values\": 5,\n        \"samples\": [\n          19955,\n          5019,\n          8534\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Name\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"0_the_of_and_in\",\n          \"3_gut_microbiota_the_and\",\n          \"1_the_of_and_to\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"CustomName\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"Metabolomics in Stress Resistance and Adaptation\",\n          \"Gut Microbiota and Human Health\",\n          \"Mass Spectrometry-Based Metabolomics Analysis\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Representation\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"GPT-4o\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Representative_Docs\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}",
       "type": "dataframe"
      },
      "text/html": [
       "\n",
       "  <div id=\"df-c126fca6-50fe-426e-a2aa-b35e6afada48\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Count</th>\n",
       "      <th>Name</th>\n",
       "      <th>CustomName</th>\n",
       "      <th>Representation</th>\n",
       "      <th>GPT-4o</th>\n",
       "      <th>Representative_Docs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "      <td>2632</td>\n",
       "      <td>-1_of_the_and_in</td>\n",
       "      <td>Metabolic Mechanisms in Disease Conditions</td>\n",
       "      <td>[of, the, and, in, to, is, with, for, that, we]</td>\n",
       "      <td>[Metabolic Mechanisms in Disease Conditions]</td>\n",
       "      <td>[Mycobacterium tuberculosis developed efficien...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>19955</td>\n",
       "      <td>0_the_of_and_in</td>\n",
       "      <td>Metabolomics in Stress Resistance and Adaptation</td>\n",
       "      <td>[the, of, and, in, to, were, that, for, as, by]</td>\n",
       "      <td>[Metabolomics in Stress Resistance and Adaptat...</td>\n",
       "      <td>[Phenolic compounds are implied in plant-micro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>8534</td>\n",
       "      <td>1_the_of_and_to</td>\n",
       "      <td>Mass Spectrometry-Based Metabolomics Analysis</td>\n",
       "      <td>[the, of, and, to, for, data, in, ms, is, mass]</td>\n",
       "      <td>[Mass Spectrometry-Based Metabolomics Analysis]</td>\n",
       "      <td>[Metabolomics, as a part of systems biology, h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>7831</td>\n",
       "      <td>2_cancer_cells_of_and</td>\n",
       "      <td>Cancer Metabolism and Biomarkers</td>\n",
       "      <td>[cancer, cells, of, and, in, the, cell, to, tu...</td>\n",
       "      <td>[Cancer Metabolism and Biomarkers]</td>\n",
       "      <td>[Rationale: It has been proposed that cancer s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>5019</td>\n",
       "      <td>3_gut_microbiota_the_and</td>\n",
       "      <td>Gut Microbiota and Human Health</td>\n",
       "      <td>[gut, microbiota, the, and, of, in, microbiome...</td>\n",
       "      <td>[Gut Microbiota and Human Health]</td>\n",
       "      <td>[Over the last five years an increasing effort...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-c126fca6-50fe-426e-a2aa-b35e6afada48')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-c126fca6-50fe-426e-a2aa-b35e6afada48 button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-c126fca6-50fe-426e-a2aa-b35e6afada48');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "<div id=\"df-3c039209-93ec-453f-b349-199943689a5b\">\n",
       "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-3c039209-93ec-453f-b349-199943689a5b')\"\n",
       "            title=\"Suggest charts\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "     width=\"24px\">\n",
       "    <g>\n",
       "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
       "    </g>\n",
       "</svg>\n",
       "  </button>\n",
       "\n",
       "<style>\n",
       "  .colab-df-quickchart {\n",
       "      --bg-color: #E8F0FE;\n",
       "      --fill-color: #1967D2;\n",
       "      --hover-bg-color: #E2EBFA;\n",
       "      --hover-fill-color: #174EA6;\n",
       "      --disabled-fill-color: #AAA;\n",
       "      --disabled-bg-color: #DDD;\n",
       "  }\n",
       "\n",
       "  [theme=dark] .colab-df-quickchart {\n",
       "      --bg-color: #3B4455;\n",
       "      --fill-color: #D2E3FC;\n",
       "      --hover-bg-color: #434B5C;\n",
       "      --hover-fill-color: #FFFFFF;\n",
       "      --disabled-bg-color: #3B4455;\n",
       "      --disabled-fill-color: #666;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart {\n",
       "    background-color: var(--bg-color);\n",
       "    border: none;\n",
       "    border-radius: 50%;\n",
       "    cursor: pointer;\n",
       "    display: none;\n",
       "    fill: var(--fill-color);\n",
       "    height: 32px;\n",
       "    padding: 0;\n",
       "    width: 32px;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart:hover {\n",
       "    background-color: var(--hover-bg-color);\n",
       "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "    fill: var(--button-hover-fill-color);\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart-complete:disabled,\n",
       "  .colab-df-quickchart-complete:disabled:hover {\n",
       "    background-color: var(--disabled-bg-color);\n",
       "    fill: var(--disabled-fill-color);\n",
       "    box-shadow: none;\n",
       "  }\n",
       "\n",
       "  .colab-df-spinner {\n",
       "    border: 2px solid var(--fill-color);\n",
       "    border-color: transparent;\n",
       "    border-bottom-color: var(--fill-color);\n",
       "    animation:\n",
       "      spin 1s steps(1) infinite;\n",
       "  }\n",
       "\n",
       "  @keyframes spin {\n",
       "    0% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "      border-left-color: var(--fill-color);\n",
       "    }\n",
       "    20% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    30% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    40% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    60% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    80% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "    90% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "  }\n",
       "</style>\n",
       "\n",
       "  <script>\n",
       "    async function quickchart(key) {\n",
       "      const quickchartButtonEl =\n",
       "        document.querySelector('#' + key + ' button');\n",
       "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
       "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
       "      try {\n",
       "        const charts = await google.colab.kernel.invokeFunction(\n",
       "            'suggestCharts', [key], {});\n",
       "      } catch (error) {\n",
       "        console.error('Error during call to suggestCharts:', error);\n",
       "      }\n",
       "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
       "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
       "    }\n",
       "    (() => {\n",
       "      let quickchartButtonEl =\n",
       "        document.querySelector('#df-3c039209-93ec-453f-b349-199943689a5b button');\n",
       "      quickchartButtonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "    })();\n",
       "  </script>\n",
       "</div>\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ],
      "text/plain": [
       "   Topic  Count                      Name  \\\n",
       "0     -1   2632          -1_of_the_and_in   \n",
       "1      0  19955           0_the_of_and_in   \n",
       "2      1   8534           1_the_of_and_to   \n",
       "3      2   7831     2_cancer_cells_of_and   \n",
       "4      3   5019  3_gut_microbiota_the_and   \n",
       "\n",
       "                                         CustomName  \\\n",
       "0        Metabolic Mechanisms in Disease Conditions   \n",
       "1  Metabolomics in Stress Resistance and Adaptation   \n",
       "2     Mass Spectrometry-Based Metabolomics Analysis   \n",
       "3                  Cancer Metabolism and Biomarkers   \n",
       "4                   Gut Microbiota and Human Health   \n",
       "\n",
       "                                      Representation  \\\n",
       "0    [of, the, and, in, to, is, with, for, that, we]   \n",
       "1    [the, of, and, in, to, were, that, for, as, by]   \n",
       "2    [the, of, and, to, for, data, in, ms, is, mass]   \n",
       "3  [cancer, cells, of, and, in, the, cell, to, tu...   \n",
       "4  [gut, microbiota, the, and, of, in, microbiome...   \n",
       "\n",
       "                                              GPT-4o  \\\n",
       "0       [Metabolic Mechanisms in Disease Conditions]   \n",
       "1  [Metabolomics in Stress Resistance and Adaptat...   \n",
       "2    [Mass Spectrometry-Based Metabolomics Analysis]   \n",
       "3                 [Cancer Metabolism and Biomarkers]   \n",
       "4                  [Gut Microbiota and Human Health]   \n",
       "\n",
       "                                 Representative_Docs  \n",
       "0  [Mycobacterium tuberculosis developed efficien...  \n",
       "1  [Phenolic compounds are implied in plant-micro...  \n",
       "2  [Metabolomics, as a part of systems biology, h...  \n",
       "3  [Rationale: It has been proposed that cancer s...  \n",
       "4  [Over the last five years an increasing effort...  "
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_model.get_topic_info().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "HDE7QECY2T0U",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### GPT-4o Mini"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "-Lxs6kz92SSp"
   },
   "outputs": [],
   "source": [
    "gpt4o_labels = [label[0][0].split(\"\\n\")[0] for label in topic_model.get_topics(full=True)[\"GPT-4o\"].values()]\n",
    "topic_model.set_topic_labels(gpt4o_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "background_save": true,
     "base_uri": "https://localhost:8080/",
     "height": 787,
     "output_embedded_package_id": "1FphjCtKueF2rzghzggvTWQT_tJbVN9jF"
    },
    "executionInfo": {
     "elapsed": 30721,
     "status": "ok",
     "timestamp": 1728500686071,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 240
    },
    "id": "6RKnJs5h2aZJ",
    "outputId": "5a524858-954a-4883-b5ab-44bf0729e1ea"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Output hidden; open in https://colab.research.google.com to view."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "topic_model.visualize_documents(titles,\n",
    "                                reduced_embeddings=reduced_embeddings,\n",
    "                                hide_annotations=True,\n",
    "                                hide_document_hover=False,\n",
    "                                custom_labels=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "LpiNSKIJJcS0",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 🗄 Backup Saving"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "7WXCwhr42ogN"
   },
   "outputs": [],
   "source": [
    "# Install safetensors if not already installed\n",
    "!pip install --quiet safetensors==0.4.5\n",
    "\n",
    "import pickle\n",
    "from bertopic import BERTopic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "t1RdNrZmJyNN"
   },
   "outputs": [],
   "source": [
    "# Save the topic model\n",
    "topic_model.save(\"topic_model\",\n",
    "                 serialization=\"safetensors\",\n",
    "                 save_ctfidf=True,\n",
    "                 #save_embedding_model=\"meta-llama/Meta-Llama-3.1-8B-Instruct\"\n",
    "                 )\n",
    "\n",
    "# Save the reduced embeddings\n",
    "with open('reduced_embeddings.pickle', 'wb') as handle:\n",
    "    pickle.dump(reduced_embeddings, handle, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "\n",
    "# Save the representative documents\n",
    "with open('rep_docs.pickle', 'wb') as handle:\n",
    "    pickle.dump(topic_model.representative_docs_,\n",
    "                handle,\n",
    "                protocol=pickle.HIGHEST_PROTOCOL)\n",
    "\n",
    "# Save Llama 3.1 labels\n",
    "#llama3_1_labels = [label[0][0].split(\"\\n\")[0] for label in topic_model.get_topics(full=True)[\"Llama3.1\"].values()]\n",
    "#with open('llama2_labels.pickle', 'wb') as handle:\n",
    "#    pickle.dump(llama3_1_labels, handle, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "\n",
    "# Save GPT-4o labels\n",
    "gpt4o_labels = [label[0][0].split(\"\\n\")[0] for label in topic_model.get_topics(full=True)[\"GPT-4o\"].values()]\n",
    "with open('gpt4o_labels.pickle', 'wb') as handle:\n",
    "    pickle.dump(gpt4o_labels, handle, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "\n",
    "# Optional: Zip the saved files\n",
    "#!zip -r /content/combined_topic_model.zip /content/combined_topic_model reduced_embeddings.pickle rep_docs.pickle llama2_labels.pickle gpt4o_labels.pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "2wO-HE6-2UUM"
   },
   "outputs": [],
   "source": [
    "# To load the saved components later:\n",
    "# loaded_model = BERTopic.load(\"combined_topic_model\")\n",
    "# with open('reduced_embeddings.pickle', 'rb') as handle:\n",
    "#     loaded_reduced_embeddings = pickle.load(handle)\n",
    "# with open('rep_docs.pickle', 'rb') as handle:\n",
    "#     loaded_rep_docs = pickle.load(handle)\n",
    "# with open('llama2_labels.pickle', 'rb') as handle:\n",
    "#     loaded_llama2_labels = pickle.load(handle)\n",
    "# with open('gpt4o_labels.pickle', 'rb') as handle:\n",
    "#     loaded_gpt4o_labels = pickle.load(handle)\n",
    "\n",
    "# To recreate visualizations:\n",
    "# topic_model.set_topic_labels(loaded_llama2_labels)  # For Llama 2 visualization\n",
    "# topic_model.visualize_documents(titles,\n",
    "#                                 reduced_embeddings=loaded_reduced_embeddings,\n",
    "#                                 hide_annotations=True,\n",
    "#                                 hide_document_hover=False,\n",
    "#                                 custom_labels=True)\n",
    "\n",
    "# topic_model.set_topic_labels(loaded_gpt4o_labels)  # For GPT-4o visualization\n",
    "# topic_model.visualize_documents(titles,\n",
    "#                                 reduced_embeddings=loaded_reduced_embeddings,\n",
    "#                                 hide_annotations=True,\n",
    "#                                 hide_document_hover=False,\n",
    "#                                 custom_labels=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "EEQXvq5X2UQ4"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "VaUTMNgW6DEp"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 62235,
     "status": "ok",
     "timestamp": 1735325657637,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 300
    },
    "id": "2hEqu6Xk60ds",
    "outputId": "d2246d54-e83c-46b9-9f91-bda54a1030ce"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading package lists...\n",
      "Building dependency tree...\n",
      "Reading state information...\n",
      "The following additional packages will be installed:\n",
      "  dvisvgm fonts-droid-fallback fonts-lato fonts-lmodern fonts-noto-mono fonts-texgyre\n",
      "  fonts-urw-base35 libapache-pom-java libcmark-gfm-extensions0.29.0.gfm.3 libcmark-gfm0.29.0.gfm.3\n",
      "  libcommons-logging-java libcommons-parent-java libfontbox-java libfontenc1 libgs9 libgs9-common\n",
      "  libidn12 libijs-0.35 libjbig2dec0 libkpathsea6 libpdfbox-java libptexenc1 libruby3.0 libsynctex2\n",
      "  libteckit0 libtexlua53 libtexluajit2 libwoff1 libzzip-0-13 lmodern pandoc-data poppler-data\n",
      "  preview-latex-style rake ruby ruby-net-telnet ruby-rubygems ruby-webrick ruby-xmlrpc ruby3.0\n",
      "  rubygems-integration t1utils teckit tex-common tex-gyre texlive-base texlive-binaries\n",
      "  texlive-fonts-recommended texlive-latex-base texlive-latex-extra texlive-latex-recommended\n",
      "  texlive-pictures texlive-plain-generic tipa xfonts-encodings xfonts-utils\n",
      "Suggested packages:\n",
      "  fonts-noto fonts-freefont-otf | fonts-freefont-ttf libavalon-framework-java\n",
      "  libcommons-logging-java-doc libexcalibur-logkit-java liblog4j1.2-java texlive-luatex\n",
      "  pandoc-citeproc context wkhtmltopdf librsvg2-bin groff ghc nodejs php python libjs-mathjax\n",
      "  libjs-katex citation-style-language-styles poppler-utils ghostscript fonts-japanese-mincho\n",
      "  | fonts-ipafont-mincho fonts-japanese-gothic | fonts-ipafont-gothic fonts-arphic-ukai\n",
      "  fonts-arphic-uming fonts-nanum ri ruby-dev bundler debhelper gv | postscript-viewer perl-tk xpdf\n",
      "  | pdf-viewer xzdec texlive-fonts-recommended-doc texlive-latex-base-doc python3-pygments\n",
      "  icc-profiles libfile-which-perl libspreadsheet-parseexcel-perl texlive-latex-extra-doc\n",
      "  texlive-latex-recommended-doc texlive-pstricks dot2tex prerex texlive-pictures-doc vprerex\n",
      "  default-jre-headless tipa-doc\n",
      "The following NEW packages will be installed:\n",
      "  dvisvgm fonts-droid-fallback fonts-lato fonts-lmodern fonts-noto-mono fonts-texgyre\n",
      "  fonts-urw-base35 libapache-pom-java libcmark-gfm-extensions0.29.0.gfm.3 libcmark-gfm0.29.0.gfm.3\n",
      "  libcommons-logging-java libcommons-parent-java libfontbox-java libfontenc1 libgs9 libgs9-common\n",
      "  libidn12 libijs-0.35 libjbig2dec0 libkpathsea6 libpdfbox-java libptexenc1 libruby3.0 libsynctex2\n",
      "  libteckit0 libtexlua53 libtexluajit2 libwoff1 libzzip-0-13 lmodern pandoc pandoc-data\n",
      "  poppler-data preview-latex-style rake ruby ruby-net-telnet ruby-rubygems ruby-webrick ruby-xmlrpc\n",
      "  ruby3.0 rubygems-integration t1utils teckit tex-common tex-gyre texlive-base texlive-binaries\n",
      "  texlive-fonts-recommended texlive-latex-base texlive-latex-extra texlive-latex-recommended\n",
      "  texlive-pictures texlive-plain-generic texlive-xetex tipa xfonts-encodings xfonts-utils\n",
      "0 upgraded, 58 newly installed, 0 to remove and 49 not upgraded.\n",
      "Need to get 202 MB of archives.\n",
      "After this operation, 728 MB of additional disk space will be used.\n",
      "Get:1 http://archive.ubuntu.com/ubuntu jammy/main amd64 fonts-droid-fallback all 1:6.0.1r16-1.1build1 [1,805 kB]\n",
      "Get:2 http://archive.ubuntu.com/ubuntu jammy/main amd64 fonts-lato all 2.0-2.1 [2,696 kB]\n",
      "Get:3 http://archive.ubuntu.com/ubuntu jammy/main amd64 poppler-data all 0.4.11-1 [2,171 kB]\n",
      "Get:4 http://archive.ubuntu.com/ubuntu jammy/universe amd64 tex-common all 6.17 [33.7 kB]\n",
      "Get:5 http://archive.ubuntu.com/ubuntu jammy/main amd64 fonts-urw-base35 all 20200910-1 [6,367 kB]\n",
      "Get:6 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libgs9-common all 9.55.0~dfsg1-0ubuntu5.10 [752 kB]\n",
      "Get:7 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libidn12 amd64 1.38-4ubuntu1 [60.0 kB]\n",
      "Get:8 http://archive.ubuntu.com/ubuntu jammy/main amd64 libijs-0.35 amd64 0.35-15build2 [16.5 kB]\n",
      "Get:9 http://archive.ubuntu.com/ubuntu jammy/main amd64 libjbig2dec0 amd64 0.19-3build2 [64.7 kB]\n",
      "Get:10 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libgs9 amd64 9.55.0~dfsg1-0ubuntu5.10 [5,031 kB]\n",
      "Get:11 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libkpathsea6 amd64 2021.20210626.59705-1ubuntu0.2 [60.4 kB]\n",
      "Get:12 http://archive.ubuntu.com/ubuntu jammy/main amd64 libwoff1 amd64 1.0.2-1build4 [45.2 kB]\n",
      "Get:13 http://archive.ubuntu.com/ubuntu jammy/universe amd64 dvisvgm amd64 2.13.1-1 [1,221 kB]\n",
      "Get:14 http://archive.ubuntu.com/ubuntu jammy/universe amd64 fonts-lmodern all 2.004.5-6.1 [4,532 kB]\n",
      "Get:15 http://archive.ubuntu.com/ubuntu jammy/main amd64 fonts-noto-mono all 20201225-1build1 [397 kB]\n",
      "Get:16 http://archive.ubuntu.com/ubuntu jammy/universe amd64 fonts-texgyre all 20180621-3.1 [10.2 MB]\n",
      "Get:17 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libapache-pom-java all 18-1 [4,720 B]\n",
      "Get:18 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcmark-gfm0.29.0.gfm.3 amd64 0.29.0.gfm.3-3 [115 kB]\n",
      "Get:19 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcmark-gfm-extensions0.29.0.gfm.3 amd64 0.29.0.gfm.3-3 [25.1 kB]\n",
      "Get:20 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-parent-java all 43-1 [10.8 kB]\n",
      "Get:21 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-logging-java all 1.2-2 [60.3 kB]\n",
      "Get:22 http://archive.ubuntu.com/ubuntu jammy/main amd64 libfontenc1 amd64 1:1.1.4-1build3 [14.7 kB]\n",
      "Get:23 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libptexenc1 amd64 2021.20210626.59705-1ubuntu0.2 [39.1 kB]\n",
      "Get:24 http://archive.ubuntu.com/ubuntu jammy/main amd64 rubygems-integration all 1.18 [5,336 B]\n",
      "Get:25 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 ruby3.0 amd64 3.0.2-7ubuntu2.8 [50.1 kB]\n",
      "Get:26 http://archive.ubuntu.com/ubuntu jammy/main amd64 ruby-rubygems all 3.3.5-2 [228 kB]\n",
      "Get:27 http://archive.ubuntu.com/ubuntu jammy/main amd64 ruby amd64 1:3.0~exp1 [5,100 B]\n",
      "Get:28 http://archive.ubuntu.com/ubuntu jammy/main amd64 rake all 13.0.6-2 [61.7 kB]\n",
      "Get:29 http://archive.ubuntu.com/ubuntu jammy/main amd64 ruby-net-telnet all 0.1.1-2 [12.6 kB]\n",
      "Get:30 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 ruby-webrick all 1.7.0-3ubuntu0.1 [52.1 kB]\n",
      "Get:31 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 ruby-xmlrpc all 0.3.2-1ubuntu0.1 [24.9 kB]\n",
      "Get:32 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libruby3.0 amd64 3.0.2-7ubuntu2.8 [5,113 kB]\n",
      "Get:33 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libsynctex2 amd64 2021.20210626.59705-1ubuntu0.2 [55.6 kB]\n",
      "Get:34 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libteckit0 amd64 2.5.11+ds1-1 [421 kB]\n",
      "Get:35 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libtexlua53 amd64 2021.20210626.59705-1ubuntu0.2 [120 kB]\n",
      "Get:36 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libtexluajit2 amd64 2021.20210626.59705-1ubuntu0.2 [267 kB]\n",
      "Get:37 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libzzip-0-13 amd64 0.13.72+dfsg.1-1.1 [27.0 kB]\n",
      "Get:38 http://archive.ubuntu.com/ubuntu jammy/main amd64 xfonts-encodings all 1:1.0.5-0ubuntu2 [578 kB]\n",
      "Get:39 http://archive.ubuntu.com/ubuntu jammy/main amd64 xfonts-utils amd64 1:7.7+6build2 [94.6 kB]\n",
      "Get:40 http://archive.ubuntu.com/ubuntu jammy/universe amd64 lmodern all 2.004.5-6.1 [9,471 kB]\n",
      "Get:41 http://archive.ubuntu.com/ubuntu jammy/universe amd64 pandoc-data all 2.9.2.1-3ubuntu2 [81.8 kB]\n",
      "Get:42 http://archive.ubuntu.com/ubuntu jammy/universe amd64 pandoc amd64 2.9.2.1-3ubuntu2 [20.3 MB]\n",
      "Get:43 http://archive.ubuntu.com/ubuntu jammy/universe amd64 preview-latex-style all 12.2-1ubuntu1 [185 kB]\n",
      "Get:44 http://archive.ubuntu.com/ubuntu jammy/main amd64 t1utils amd64 1.41-4build2 [61.3 kB]\n",
      "Get:45 http://archive.ubuntu.com/ubuntu jammy/universe amd64 teckit amd64 2.5.11+ds1-1 [699 kB]\n",
      "Get:46 http://archive.ubuntu.com/ubuntu jammy/universe amd64 tex-gyre all 20180621-3.1 [6,209 kB]\n",
      "Get:47 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 texlive-binaries amd64 2021.20210626.59705-1ubuntu0.2 [9,860 kB]\n",
      "Get:48 http://archive.ubuntu.com/ubuntu jammy/universe amd64 texlive-base all 2021.20220204-1 [21.0 MB]\n",
      "Get:49 http://archive.ubuntu.com/ubuntu jammy/universe amd64 texlive-fonts-recommended all 2021.20220204-1 [4,972 kB]\n",
      "Get:50 http://archive.ubuntu.com/ubuntu jammy/universe amd64 texlive-latex-base all 2021.20220204-1 [1,128 kB]\n",
      "Get:51 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libfontbox-java all 1:1.8.16-2 [207 kB]\n",
      "Get:52 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libpdfbox-java all 1:1.8.16-2 [5,199 kB]\n",
      "Get:53 http://archive.ubuntu.com/ubuntu jammy/universe amd64 texlive-latex-recommended all 2021.20220204-1 [14.4 MB]\n",
      "Get:54 http://archive.ubuntu.com/ubuntu jammy/universe amd64 texlive-pictures all 2021.20220204-1 [8,720 kB]\n",
      "Get:55 http://archive.ubuntu.com/ubuntu jammy/universe amd64 texlive-latex-extra all 2021.20220204-1 [13.9 MB]\n",
      "Get:56 http://archive.ubuntu.com/ubuntu jammy/universe amd64 texlive-plain-generic all 2021.20220204-1 [27.5 MB]\n",
      "Get:57 http://archive.ubuntu.com/ubuntu jammy/universe amd64 tipa all 2:1.3-21 [2,967 kB]\n",
      "Get:58 http://archive.ubuntu.com/ubuntu jammy/universe amd64 texlive-xetex all 2021.20220204-1 [12.4 MB]\n",
      "Fetched 202 MB in 8s (24.4 MB/s)\n",
      "Extracting templates from packages: 100%\n",
      "Preconfiguring packages ...\n",
      "Selecting previously unselected package fonts-droid-fallback.\n",
      "(Reading database ... 123634 files and directories currently installed.)\n",
      "Preparing to unpack .../00-fonts-droid-fallback_1%3a6.0.1r16-1.1build1_all.deb ...\n",
      "Unpacking fonts-droid-fallback (1:6.0.1r16-1.1build1) ...\n",
      "Selecting previously unselected package fonts-lato.\n",
      "Preparing to unpack .../01-fonts-lato_2.0-2.1_all.deb ...\n",
      "Unpacking fonts-lato (2.0-2.1) ...\n",
      "Selecting previously unselected package poppler-data.\n",
      "Preparing to unpack .../02-poppler-data_0.4.11-1_all.deb ...\n",
      "Unpacking poppler-data (0.4.11-1) ...\n",
      "Selecting previously unselected package tex-common.\n",
      "Preparing to unpack .../03-tex-common_6.17_all.deb ...\n",
      "Unpacking tex-common (6.17) ...\n",
      "Selecting previously unselected package fonts-urw-base35.\n",
      "Preparing to unpack .../04-fonts-urw-base35_20200910-1_all.deb ...\n",
      "Unpacking fonts-urw-base35 (20200910-1) ...\n",
      "Selecting previously unselected package libgs9-common.\n",
      "Preparing to unpack .../05-libgs9-common_9.55.0~dfsg1-0ubuntu5.10_all.deb ...\n",
      "Unpacking libgs9-common (9.55.0~dfsg1-0ubuntu5.10) ...\n",
      "Selecting previously unselected package libidn12:amd64.\n",
      "Preparing to unpack .../06-libidn12_1.38-4ubuntu1_amd64.deb ...\n",
      "Unpacking libidn12:amd64 (1.38-4ubuntu1) ...\n",
      "Selecting previously unselected package libijs-0.35:amd64.\n",
      "Preparing to unpack .../07-libijs-0.35_0.35-15build2_amd64.deb ...\n",
      "Unpacking libijs-0.35:amd64 (0.35-15build2) ...\n",
      "Selecting previously unselected package libjbig2dec0:amd64.\n",
      "Preparing to unpack .../08-libjbig2dec0_0.19-3build2_amd64.deb ...\n",
      "Unpacking libjbig2dec0:amd64 (0.19-3build2) ...\n",
      "Selecting previously unselected package libgs9:amd64.\n",
      "Preparing to unpack .../09-libgs9_9.55.0~dfsg1-0ubuntu5.10_amd64.deb ...\n",
      "Unpacking libgs9:amd64 (9.55.0~dfsg1-0ubuntu5.10) ...\n",
      "Selecting previously unselected package libkpathsea6:amd64.\n",
      "Preparing to unpack .../10-libkpathsea6_2021.20210626.59705-1ubuntu0.2_amd64.deb ...\n",
      "Unpacking libkpathsea6:amd64 (2021.20210626.59705-1ubuntu0.2) ...\n",
      "Selecting previously unselected package libwoff1:amd64.\n",
      "Preparing to unpack .../11-libwoff1_1.0.2-1build4_amd64.deb ...\n",
      "Unpacking libwoff1:amd64 (1.0.2-1build4) ...\n",
      "Selecting previously unselected package dvisvgm.\n",
      "Preparing to unpack .../12-dvisvgm_2.13.1-1_amd64.deb ...\n",
      "Unpacking dvisvgm (2.13.1-1) ...\n",
      "Selecting previously unselected package fonts-lmodern.\n",
      "Preparing to unpack .../13-fonts-lmodern_2.004.5-6.1_all.deb ...\n",
      "Unpacking fonts-lmodern (2.004.5-6.1) ...\n",
      "Selecting previously unselected package fonts-noto-mono.\n",
      "Preparing to unpack .../14-fonts-noto-mono_20201225-1build1_all.deb ...\n",
      "Unpacking fonts-noto-mono (20201225-1build1) ...\n",
      "Selecting previously unselected package fonts-texgyre.\n",
      "Preparing to unpack .../15-fonts-texgyre_20180621-3.1_all.deb ...\n",
      "Unpacking fonts-texgyre (20180621-3.1) ...\n",
      "Selecting previously unselected package libapache-pom-java.\n",
      "Preparing to unpack .../16-libapache-pom-java_18-1_all.deb ...\n",
      "Unpacking libapache-pom-java (18-1) ...\n",
      "Selecting previously unselected package libcmark-gfm0.29.0.gfm.3:amd64.\n",
      "Preparing to unpack .../17-libcmark-gfm0.29.0.gfm.3_0.29.0.gfm.3-3_amd64.deb ...\n",
      "Unpacking libcmark-gfm0.29.0.gfm.3:amd64 (0.29.0.gfm.3-3) ...\n",
      "Selecting previously unselected package libcmark-gfm-extensions0.29.0.gfm.3:amd64.\n",
      "Preparing to unpack .../18-libcmark-gfm-extensions0.29.0.gfm.3_0.29.0.gfm.3-3_amd64.deb ...\n",
      "Unpacking libcmark-gfm-extensions0.29.0.gfm.3:amd64 (0.29.0.gfm.3-3) ...\n",
      "Selecting previously unselected package libcommons-parent-java.\n",
      "Preparing to unpack .../19-libcommons-parent-java_43-1_all.deb ...\n",
      "Unpacking libcommons-parent-java (43-1) ...\n",
      "Selecting previously unselected package libcommons-logging-java.\n",
      "Preparing to unpack .../20-libcommons-logging-java_1.2-2_all.deb ...\n",
      "Unpacking libcommons-logging-java (1.2-2) ...\n",
      "Selecting previously unselected package libfontenc1:amd64.\n",
      "Preparing to unpack .../21-libfontenc1_1%3a1.1.4-1build3_amd64.deb ...\n",
      "Unpacking libfontenc1:amd64 (1:1.1.4-1build3) ...\n",
      "Selecting previously unselected package libptexenc1:amd64.\n",
      "Preparing to unpack .../22-libptexenc1_2021.20210626.59705-1ubuntu0.2_amd64.deb ...\n",
      "Unpacking libptexenc1:amd64 (2021.20210626.59705-1ubuntu0.2) ...\n",
      "Selecting previously unselected package rubygems-integration.\n",
      "Preparing to unpack .../23-rubygems-integration_1.18_all.deb ...\n",
      "Unpacking rubygems-integration (1.18) ...\n",
      "Selecting previously unselected package ruby3.0.\n",
      "Preparing to unpack .../24-ruby3.0_3.0.2-7ubuntu2.8_amd64.deb ...\n",
      "Unpacking ruby3.0 (3.0.2-7ubuntu2.8) ...\n",
      "Selecting previously unselected package ruby-rubygems.\n",
      "Preparing to unpack .../25-ruby-rubygems_3.3.5-2_all.deb ...\n",
      "Unpacking ruby-rubygems (3.3.5-2) ...\n",
      "Selecting previously unselected package ruby.\n",
      "Preparing to unpack .../26-ruby_1%3a3.0~exp1_amd64.deb ...\n",
      "Unpacking ruby (1:3.0~exp1) ...\n",
      "Selecting previously unselected package rake.\n",
      "Preparing to unpack .../27-rake_13.0.6-2_all.deb ...\n",
      "Unpacking rake (13.0.6-2) ...\n",
      "Selecting previously unselected package ruby-net-telnet.\n",
      "Preparing to unpack .../28-ruby-net-telnet_0.1.1-2_all.deb ...\n",
      "Unpacking ruby-net-telnet (0.1.1-2) ...\n",
      "Selecting previously unselected package ruby-webrick.\n",
      "Preparing to unpack .../29-ruby-webrick_1.7.0-3ubuntu0.1_all.deb ...\n",
      "Unpacking ruby-webrick (1.7.0-3ubuntu0.1) ...\n",
      "Selecting previously unselected package ruby-xmlrpc.\n",
      "Preparing to unpack .../30-ruby-xmlrpc_0.3.2-1ubuntu0.1_all.deb ...\n",
      "Unpacking ruby-xmlrpc (0.3.2-1ubuntu0.1) ...\n",
      "Selecting previously unselected package libruby3.0:amd64.\n",
      "Preparing to unpack .../31-libruby3.0_3.0.2-7ubuntu2.8_amd64.deb ...\n",
      "Unpacking libruby3.0:amd64 (3.0.2-7ubuntu2.8) ...\n",
      "Selecting previously unselected package libsynctex2:amd64.\n",
      "Preparing to unpack .../32-libsynctex2_2021.20210626.59705-1ubuntu0.2_amd64.deb ...\n",
      "Unpacking libsynctex2:amd64 (2021.20210626.59705-1ubuntu0.2) ...\n",
      "Selecting previously unselected package libteckit0:amd64.\n",
      "Preparing to unpack .../33-libteckit0_2.5.11+ds1-1_amd64.deb ...\n",
      "Unpacking libteckit0:amd64 (2.5.11+ds1-1) ...\n",
      "Selecting previously unselected package libtexlua53:amd64.\n",
      "Preparing to unpack .../34-libtexlua53_2021.20210626.59705-1ubuntu0.2_amd64.deb ...\n",
      "Unpacking libtexlua53:amd64 (2021.20210626.59705-1ubuntu0.2) ...\n",
      "Selecting previously unselected package libtexluajit2:amd64.\n",
      "Preparing to unpack .../35-libtexluajit2_2021.20210626.59705-1ubuntu0.2_amd64.deb ...\n",
      "Unpacking libtexluajit2:amd64 (2021.20210626.59705-1ubuntu0.2) ...\n",
      "Selecting previously unselected package libzzip-0-13:amd64.\n",
      "Preparing to unpack .../36-libzzip-0-13_0.13.72+dfsg.1-1.1_amd64.deb ...\n",
      "Unpacking libzzip-0-13:amd64 (0.13.72+dfsg.1-1.1) ...\n",
      "Selecting previously unselected package xfonts-encodings.\n",
      "Preparing to unpack .../37-xfonts-encodings_1%3a1.0.5-0ubuntu2_all.deb ...\n",
      "Unpacking xfonts-encodings (1:1.0.5-0ubuntu2) ...\n",
      "Selecting previously unselected package xfonts-utils.\n",
      "Preparing to unpack .../38-xfonts-utils_1%3a7.7+6build2_amd64.deb ...\n",
      "Unpacking xfonts-utils (1:7.7+6build2) ...\n",
      "Selecting previously unselected package lmodern.\n",
      "Preparing to unpack .../39-lmodern_2.004.5-6.1_all.deb ...\n",
      "Unpacking lmodern (2.004.5-6.1) ...\n",
      "Selecting previously unselected package pandoc-data.\n",
      "Preparing to unpack .../40-pandoc-data_2.9.2.1-3ubuntu2_all.deb ...\n",
      "Unpacking pandoc-data (2.9.2.1-3ubuntu2) ...\n",
      "Selecting previously unselected package pandoc.\n",
      "Preparing to unpack .../41-pandoc_2.9.2.1-3ubuntu2_amd64.deb ...\n",
      "Unpacking pandoc (2.9.2.1-3ubuntu2) ...\n",
      "Selecting previously unselected package preview-latex-style.\n",
      "Preparing to unpack .../42-preview-latex-style_12.2-1ubuntu1_all.deb ...\n",
      "Unpacking preview-latex-style (12.2-1ubuntu1) ...\n",
      "Selecting previously unselected package t1utils.\n",
      "Preparing to unpack .../43-t1utils_1.41-4build2_amd64.deb ...\n",
      "Unpacking t1utils (1.41-4build2) ...\n",
      "Selecting previously unselected package teckit.\n",
      "Preparing to unpack .../44-teckit_2.5.11+ds1-1_amd64.deb ...\n",
      "Unpacking teckit (2.5.11+ds1-1) ...\n",
      "Selecting previously unselected package tex-gyre.\n",
      "Preparing to unpack .../45-tex-gyre_20180621-3.1_all.deb ...\n",
      "Unpacking tex-gyre (20180621-3.1) ...\n",
      "Selecting previously unselected package texlive-binaries.\n",
      "Preparing to unpack .../46-texlive-binaries_2021.20210626.59705-1ubuntu0.2_amd64.deb ...\n",
      "Unpacking texlive-binaries (2021.20210626.59705-1ubuntu0.2) ...\n",
      "Selecting previously unselected package texlive-base.\n",
      "Preparing to unpack .../47-texlive-base_2021.20220204-1_all.deb ...\n",
      "Unpacking texlive-base (2021.20220204-1) ...\n",
      "Selecting previously unselected package texlive-fonts-recommended.\n",
      "Preparing to unpack .../48-texlive-fonts-recommended_2021.20220204-1_all.deb ...\n",
      "Unpacking texlive-fonts-recommended (2021.20220204-1) ...\n",
      "Selecting previously unselected package texlive-latex-base.\n",
      "Preparing to unpack .../49-texlive-latex-base_2021.20220204-1_all.deb ...\n",
      "Unpacking texlive-latex-base (2021.20220204-1) ...\n",
      "Selecting previously unselected package libfontbox-java.\n",
      "Preparing to unpack .../50-libfontbox-java_1%3a1.8.16-2_all.deb ...\n",
      "Unpacking libfontbox-java (1:1.8.16-2) ...\n",
      "Selecting previously unselected package libpdfbox-java.\n",
      "Preparing to unpack .../51-libpdfbox-java_1%3a1.8.16-2_all.deb ...\n",
      "Unpacking libpdfbox-java (1:1.8.16-2) ...\n",
      "Selecting previously unselected package texlive-latex-recommended.\n",
      "Preparing to unpack .../52-texlive-latex-recommended_2021.20220204-1_all.deb ...\n",
      "Unpacking texlive-latex-recommended (2021.20220204-1) ...\n",
      "Selecting previously unselected package texlive-pictures.\n",
      "Preparing to unpack .../53-texlive-pictures_2021.20220204-1_all.deb ...\n",
      "Unpacking texlive-pictures (2021.20220204-1) ...\n",
      "Selecting previously unselected package texlive-latex-extra.\n",
      "Preparing to unpack .../54-texlive-latex-extra_2021.20220204-1_all.deb ...\n",
      "Unpacking texlive-latex-extra (2021.20220204-1) ...\n",
      "Selecting previously unselected package texlive-plain-generic.\n",
      "Preparing to unpack .../55-texlive-plain-generic_2021.20220204-1_all.deb ...\n",
      "Unpacking texlive-plain-generic (2021.20220204-1) ...\n",
      "Selecting previously unselected package tipa.\n",
      "Preparing to unpack .../56-tipa_2%3a1.3-21_all.deb ...\n",
      "Unpacking tipa (2:1.3-21) ...\n",
      "Selecting previously unselected package texlive-xetex.\n",
      "Preparing to unpack .../57-texlive-xetex_2021.20220204-1_all.deb ...\n",
      "Unpacking texlive-xetex (2021.20220204-1) ...\n",
      "Setting up fonts-lato (2.0-2.1) ...\n",
      "Setting up fonts-noto-mono (20201225-1build1) ...\n",
      "Setting up libwoff1:amd64 (1.0.2-1build4) ...\n",
      "Setting up libtexlua53:amd64 (2021.20210626.59705-1ubuntu0.2) ...\n",
      "Setting up libijs-0.35:amd64 (0.35-15build2) ...\n",
      "Setting up libtexluajit2:amd64 (2021.20210626.59705-1ubuntu0.2) ...\n",
      "Setting up libfontbox-java (1:1.8.16-2) ...\n",
      "Setting up rubygems-integration (1.18) ...\n",
      "Setting up libzzip-0-13:amd64 (0.13.72+dfsg.1-1.1) ...\n",
      "Setting up fonts-urw-base35 (20200910-1) ...\n",
      "Setting up poppler-data (0.4.11-1) ...\n",
      "Setting up tex-common (6.17) ...\n",
      "update-language: texlive-base not installed and configured, doing nothing!\n",
      "Setting up libfontenc1:amd64 (1:1.1.4-1build3) ...\n",
      "Setting up libjbig2dec0:amd64 (0.19-3build2) ...\n",
      "Setting up libteckit0:amd64 (2.5.11+ds1-1) ...\n",
      "Setting up libapache-pom-java (18-1) ...\n",
      "Setting up ruby-net-telnet (0.1.1-2) ...\n",
      "Setting up xfonts-encodings (1:1.0.5-0ubuntu2) ...\n",
      "Setting up t1utils (1.41-4build2) ...\n",
      "Setting up libidn12:amd64 (1.38-4ubuntu1) ...\n",
      "Setting up fonts-texgyre (20180621-3.1) ...\n",
      "Setting up libkpathsea6:amd64 (2021.20210626.59705-1ubuntu0.2) ...\n",
      "Setting up ruby-webrick (1.7.0-3ubuntu0.1) ...\n",
      "Setting up libcmark-gfm0.29.0.gfm.3:amd64 (0.29.0.gfm.3-3) ...\n",
      "Setting up fonts-lmodern (2.004.5-6.1) ...\n",
      "Setting up libcmark-gfm-extensions0.29.0.gfm.3:amd64 (0.29.0.gfm.3-3) ...\n",
      "Setting up fonts-droid-fallback (1:6.0.1r16-1.1build1) ...\n",
      "Setting up pandoc-data (2.9.2.1-3ubuntu2) ...\n",
      "Setting up ruby-xmlrpc (0.3.2-1ubuntu0.1) ...\n",
      "Setting up libsynctex2:amd64 (2021.20210626.59705-1ubuntu0.2) ...\n",
      "Setting up libgs9-common (9.55.0~dfsg1-0ubuntu5.10) ...\n",
      "Setting up teckit (2.5.11+ds1-1) ...\n",
      "Setting up libpdfbox-java (1:1.8.16-2) ...\n",
      "Setting up libgs9:amd64 (9.55.0~dfsg1-0ubuntu5.10) ...\n",
      "Setting up preview-latex-style (12.2-1ubuntu1) ...\n",
      "Setting up libcommons-parent-java (43-1) ...\n",
      "Setting up dvisvgm (2.13.1-1) ...\n",
      "Setting up libcommons-logging-java (1.2-2) ...\n",
      "Setting up xfonts-utils (1:7.7+6build2) ...\n",
      "Setting up libptexenc1:amd64 (2021.20210626.59705-1ubuntu0.2) ...\n",
      "Setting up pandoc (2.9.2.1-3ubuntu2) ...\n",
      "Setting up texlive-binaries (2021.20210626.59705-1ubuntu0.2) ...\n",
      "update-alternatives: using /usr/bin/xdvi-xaw to provide /usr/bin/xdvi.bin (xdvi.bin) in auto mode\n",
      "update-alternatives: using /usr/bin/bibtex.original to provide /usr/bin/bibtex (bibtex) in auto mode\n",
      "Setting up lmodern (2.004.5-6.1) ...\n",
      "Setting up texlive-base (2021.20220204-1) ...\n",
      "/usr/bin/ucfr\n",
      "/usr/bin/ucfr\n",
      "/usr/bin/ucfr\n",
      "/usr/bin/ucfr\n",
      "mktexlsr: Updating /var/lib/texmf/ls-R-TEXLIVEDIST... \n",
      "mktexlsr: Updating /var/lib/texmf/ls-R-TEXMFMAIN... \n",
      "mktexlsr: Updating /var/lib/texmf/ls-R... \n",
      "mktexlsr: Done.\n",
      "tl-paper: setting paper size for dvips to a4: /var/lib/texmf/dvips/config/config-paper.ps\n",
      "tl-paper: setting paper size for dvipdfmx to a4: /var/lib/texmf/dvipdfmx/dvipdfmx-paper.cfg\n",
      "tl-paper: setting paper size for xdvi to a4: /var/lib/texmf/xdvi/XDvi-paper\n",
      "tl-paper: setting paper size for pdftex to a4: /var/lib/texmf/tex/generic/tex-ini-files/pdftexconfig.tex\n",
      "Setting up tex-gyre (20180621-3.1) ...\n",
      "Setting up texlive-plain-generic (2021.20220204-1) ...\n",
      "Setting up texlive-latex-base (2021.20220204-1) ...\n",
      "Setting up texlive-latex-recommended (2021.20220204-1) ...\n",
      "Setting up texlive-pictures (2021.20220204-1) ...\n",
      "Setting up texlive-fonts-recommended (2021.20220204-1) ...\n",
      "Setting up tipa (2:1.3-21) ...\n",
      "Setting up texlive-latex-extra (2021.20220204-1) ...\n",
      "Setting up texlive-xetex (2021.20220204-1) ...\n",
      "Setting up rake (13.0.6-2) ...\n",
      "Setting up libruby3.0:amd64 (3.0.2-7ubuntu2.8) ...\n",
      "Setting up ruby3.0 (3.0.2-7ubuntu2.8) ...\n",
      "Setting up ruby (1:3.0~exp1) ...\n",
      "Setting up ruby-rubygems (3.3.5-2) ...\n",
      "Processing triggers for man-db (2.10.2-1) ...\n",
      "Processing triggers for fontconfig (2.13.1-4.2ubuntu5) ...\n",
      "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
      "\n",
      "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
      "\n",
      "Processing triggers for tex-common (6.17) ...\n",
      "Running updmap-sys. This may take some time... done.\n",
      "Running mktexlsr /var/lib/texmf ... done.\n",
      "Building format(s) --all.\n",
      "\tThis may take some time... done.\n"
     ]
    }
   ],
   "source": [
    "!apt-get install --quiet texlive-xetex pandoc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 10733,
     "status": "ok",
     "timestamp": 1735325697290,
     "user": {
      "displayName": "Olatomiwa Bifarin",
      "userId": "06174025613084979381"
     },
     "user_tz": 300
    },
    "id": "yT85E-YKdPYy",
    "outputId": "4a435f4d-8026-4360-91f2-f9001f541b6b"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook topic-modelling.ipynb to pdf\n",
      "/usr/local/share/jupyter/nbconvert/templates/latex/display_priority.j2:32: UserWarning: Your element with mimetype(s) dict_keys(['text/html']) is not able to be represented.\n",
      "  ((*- endblock -*))\n",
      "[NbConvertApp] Writing 104052 bytes to notebook.tex\n",
      "[NbConvertApp] Building PDF\n",
      "[NbConvertApp] Running xelatex 3 times: ['xelatex', 'notebook.tex', '-quiet']\n",
      "[NbConvertApp] Running bibtex 1 time: ['bibtex', 'notebook']\n",
      "[NbConvertApp] WARNING | bibtex had problems, most likely because there were no citations\n",
      "[NbConvertApp] PDF successfully created\n",
      "[NbConvertApp] Writing 126592 bytes to topic-modelling.pdf\n"
     ]
    }
   ],
   "source": [
    "#README: you need to copy this notebook into the folder session first for this code to work.\n",
    "import os\n",
    "os.chdir('/content')\n",
    "!jupyter nbconvert --to pdf \"topic-modelling.ipynb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "SDyud-9udPUw"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "authorship_tag": "ABX9TyMxLjixP5sQqinpOji2o+EK",
   "collapsed_sections": [
    "dvduAcVTLtgY",
    "y7jjL_5LL2_N",
    "yxhAcmuGMYYW",
    "-Qs1Rxw2Mm9I",
    "bMLNM2O_NCcO",
    "6LKyccJidu3K",
    "jl-WT_1Gd466",
    "-SzWRahc-NnP",
    "_lJ-wcd47x5W",
    "RVyAstZtf3TP",
    "9gZfgW8o4-tl",
    "vWemkaDIepqm",
    "fqkWKW7CfF9M",
    "BpOmV4Dmjtt9",
    "1rCuEpGXf8WR",
    "HDE7QECY2T0U",
    "oyCfWgLm2N_p",
    "LpiNSKIJJcS0"
   ],
   "gpuType": "T4",
   "machine_shape": "hm",
   "provenance": [],
   "toc_visible": true
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "169fed61ddbb4f96a9a178eda9d3af5d": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "175db7c0169646cc8c340cb10df6b68f": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "ProgressStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "ProgressStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "bar_color": null,
      "description_width": ""
     }
    },
    "3c68485af6724380ac269f1747e4b479": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "84754e2366ab4bc1b9f1b0c6470f4879": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "FloatProgressModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "FloatProgressModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "ProgressView",
      "bar_style": "success",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_ffb91fa10ae6449fbfb89113ee16062d",
      "max": 2521,
      "min": 0,
      "orientation": "horizontal",
      "style": "IPY_MODEL_175db7c0169646cc8c340cb10df6b68f",
      "value": 2521
     }
    },
    "8e520224d0324e139a9b4528639e4d75": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_3c68485af6724380ac269f1747e4b479",
      "placeholder": "​",
      "style": "IPY_MODEL_d94785a14eec4a6ab914f1691b4cc811",
      "value": "Batches: 100%"
     }
    },
    "95fd75e407bd45ba9803b8d347f99173": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "ca327b5c507b4dc28b7400ea0df30b66": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_95fd75e407bd45ba9803b8d347f99173",
      "placeholder": "​",
      "style": "IPY_MODEL_169fed61ddbb4f96a9a178eda9d3af5d",
      "value": " 2521/2521 [03:19&lt;00:00, 69.84it/s]"
     }
    },
    "d94785a14eec4a6ab914f1691b4cc811": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "e08b89598596417ea2183100097f33be": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HBoxModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HBoxModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HBoxView",
      "box_style": "",
      "children": [
       "IPY_MODEL_8e520224d0324e139a9b4528639e4d75",
       "IPY_MODEL_84754e2366ab4bc1b9f1b0c6470f4879",
       "IPY_MODEL_ca327b5c507b4dc28b7400ea0df30b66"
      ],
      "layout": "IPY_MODEL_ef49e2bf2d44470da616ae07ddb270a9"
     }
    },
    "ef49e2bf2d44470da616ae07ddb270a9": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "ffb91fa10ae6449fbfb89113ee16062d": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
